{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Learning sgRNA predictors from empirical data\n",
    "    * Load scripts and empirical data\n",
    "    * Generate TSS annotation using FANTOM dataset\n",
    "    * Calculate parameters for empirical sgRNAs\n",
    "    * Fit parameters\n",
    "2. Applying machine learning model to predict sgRNA activity\n",
    "    * Find all sgRNAs in genomic regions of interest \n",
    "    * Predicting sgRNA activity\n",
    "3. Construct sgRNA libraries\n",
    "    * Score sgRNAs for off-target potential\n",
    "* Pick the top sgRNAs for a library, given predicted activity scores and off-target filtering\n",
    "* Design negative controls matching the base composition of the library\n",
    "* Finalizing library design"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make sure to run `jupyter notebook` after running `source activate CRISPRiaDesign`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Learning sgRNA predictors from empirical data\n",
    "## Load scripts and empirical data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ucbbiosciences/software/anaconda2/envs/CRISPRiaDesign/lib/python2.7/site-packages/sklearn/ensemble/weight_boosting.py:29: DeprecationWarning: numpy.core.umath_tests is an internal NumPy module and should not be imported. It will be removed in a future NumPy release.\n",
      "  from numpy.core.umath_tests import inner1d\n",
      "/Users/ucbbiosciences/software/anaconda2/envs/CRISPRiaDesign/lib/python2.7/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.\n",
      "  \"This module will be removed in 0.20.\", DeprecationWarning)\n",
      "/Users/ucbbiosciences/software/anaconda2/envs/CRISPRiaDesign/lib/python2.7/site-packages/sklearn/grid_search.py:42: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. This module will be removed in 0.20.\n",
      "  DeprecationWarning)\n"
     ]
    }
   ],
   "source": [
    "%run sgRNA_learning.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading genome file...Done\n"
     ]
    }
   ],
   "source": [
    "genomeDict = loadGenomeAsDict(os.path.expanduser('~/data/reference-data-management/genomes/hg19.fa'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "OrderedDict([('chr1',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chr1', name='chr1', description='chr1 dna:chromosome chromosome:GRCh37:1:1:249250621:1', dbxrefs=[])),\n",
       "             ('chr2',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chr2', name='chr2', description='chr2 dna:chromosome chromosome:GRCh37:2:1:243199373:1', dbxrefs=[])),\n",
       "             ('chr3',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chr3', name='chr3', description='chr3 dna:chromosome chromosome:GRCh37:3:1:198022430:1', dbxrefs=[])),\n",
       "             ('chr4',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chr4', name='chr4', description='chr4 dna:chromosome chromosome:GRCh37:4:1:191154276:1', dbxrefs=[])),\n",
       "             ('chr5',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chr5', name='chr5', description='chr5 dna:chromosome chromosome:GRCh37:5:1:180915260:1', dbxrefs=[])),\n",
       "             ('chr6',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chr6', name='chr6', description='chr6 dna:chromosome chromosome:GRCh37:6:1:171115067:1', dbxrefs=[])),\n",
       "             ('chr7',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chr7', name='chr7', description='chr7 dna:chromosome chromosome:GRCh37:7:1:159138663:1', dbxrefs=[])),\n",
       "             ('chr8',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chr8', name='chr8', description='chr8 dna:chromosome chromosome:GRCh37:8:1:146364022:1', dbxrefs=[])),\n",
       "             ('chr9',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chr9', name='chr9', description='chr9 dna:chromosome chromosome:GRCh37:9:1:141213431:1', dbxrefs=[])),\n",
       "             ('chr10',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chr10', name='chr10', description='chr10 dna:chromosome chromosome:GRCh37:10:1:135534747:1', dbxrefs=[])),\n",
       "             ('chr11',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chr11', name='chr11', description='chr11 dna:chromosome chromosome:GRCh37:11:1:135006516:1', dbxrefs=[])),\n",
       "             ('chr12',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chr12', name='chr12', description='chr12 dna:chromosome chromosome:GRCh37:12:1:133851895:1', dbxrefs=[])),\n",
       "             ('chr13',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chr13', name='chr13', description='chr13 dna:chromosome chromosome:GRCh37:13:1:115169878:1', dbxrefs=[])),\n",
       "             ('chr14',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chr14', name='chr14', description='chr14 dna:chromosome chromosome:GRCh37:14:1:107349540:1', dbxrefs=[])),\n",
       "             ('chr15',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chr15', name='chr15', description='chr15 dna:chromosome chromosome:GRCh37:15:1:102531392:1', dbxrefs=[])),\n",
       "             ('chr16',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chr16', name='chr16', description='chr16 dna:chromosome chromosome:GRCh37:16:1:90354753:1', dbxrefs=[])),\n",
       "             ('chr17',\n",
       "              SeqRecord(seq=Seq('AAGCTTCTCACCCTGTTCCTGCATAGATAATTGCATGACAATTGCCTTGTCCCT...GGT', SingleLetterAlphabet()), id='chr17', name='chr17', description='chr17 dna:chromosome chromosome:GRCh37:17:1:81195210:1', dbxrefs=[])),\n",
       "             ('chr18',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chr18', name='chr18', description='chr18 dna:chromosome chromosome:GRCh37:18:1:78077248:1', dbxrefs=[])),\n",
       "             ('chr19',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chr19', name='chr19', description='chr19 dna:chromosome chromosome:GRCh37:19:1:59128983:1', dbxrefs=[])),\n",
       "             ('chr20',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chr20', name='chr20', description='chr20 dna:chromosome chromosome:GRCh37:20:1:63025520:1', dbxrefs=[])),\n",
       "             ('chr21',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chr21', name='chr21', description='chr21 dna:chromosome chromosome:GRCh37:21:1:48129895:1', dbxrefs=[])),\n",
       "             ('chr22',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chr22', name='chr22', description='chr22 dna:chromosome chromosome:GRCh37:22:1:51304566:1', dbxrefs=[])),\n",
       "             ('chrX',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chrX', name='chrX', description='chrX dna:chromosome chromosome:GRCh37:X:1:155270560:1', dbxrefs=[])),\n",
       "             ('chrY',\n",
       "              SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', SingleLetterAlphabet()), id='chrY', name='chrY', description='chrY dna:chromosome chromosome:GRCh37:Y:2649521:59034049:1', dbxrefs=[])),\n",
       "             ('chrMT',\n",
       "              SeqRecord(seq=Seq('GATCACAGGTCTATCACCCTATTAACCACTCACGGGAGCTCTCCATGCATTTGG...ATG', SingleLetterAlphabet()), id='chrMT', name='chrMT', description='chrMT gi|251831106|ref|NC_012920.1| Homo sapiens mitochondrion, complete genome', dbxrefs=[])),\n",
       "             ('chrGL000207.1',\n",
       "              SeqRecord(seq=Seq('CATACATTTAATATACCCTCACCATACAGAATGTTCTTTCCCTATTACATAAGG...ATC', SingleLetterAlphabet()), id='chrGL000207.1', name='chrGL000207.1', description='chrGL000207.1 dna:supercontig supercontig::GL000207.1:1:4262:1', dbxrefs=[])),\n",
       "             ('chrGL000226.1',\n",
       "              SeqRecord(seq=Seq('GATCTTTCTTTTGATAGAGCAGTTCTGAAAAACACTTTTTGTTGAATCTGCAAG...ATC', SingleLetterAlphabet()), id='chrGL000226.1', name='chrGL000226.1', description='chrGL000226.1 dna:supercontig supercontig::GL000226.1:1:15008:1', dbxrefs=[])),\n",
       "             ('chrGL000229.1',\n",
       "              SeqRecord(seq=Seq('GCCGCTGTGGGATTTTCCGTGGGAGAGGTATGGGAGAGGACTGACGCGGCTTCC...ATC', SingleLetterAlphabet()), id='chrGL000229.1', name='chrGL000229.1', description='chrGL000229.1 dna:supercontig supercontig::GL000229.1:1:19913:1', dbxrefs=[])),\n",
       "             ('chrGL000231.1',\n",
       "              SeqRecord(seq=Seq('TTTAATGAATAAGTATTTACGCTATTAATATTTATAATATTGGTGCTATTATTT...ATC', SingleLetterAlphabet()), id='chrGL000231.1', name='chrGL000231.1', description='chrGL000231.1 dna:supercontig supercontig::GL000231.1:1:27386:1', dbxrefs=[])),\n",
       "             ('chrGL000210.1',\n",
       "              SeqRecord(seq=Seq('AAAAGAACAGCTGGCCTGGGCTTGCTGCTGGGATGCGGGTGGGGCTCTGGGGGG...TCA', SingleLetterAlphabet()), id='chrGL000210.1', name='chrGL000210.1', description='chrGL000210.1 dna:supercontig supercontig::GL000210.1:1:27682:1', dbxrefs=[])),\n",
       "             ('chrGL000239.1',\n",
       "              SeqRecord(seq=Seq('GATCTCTCTATCGGCCTGCTCTGGGCCACAGAGACTGGCCCTGACGACTTCTCA...ATC', SingleLetterAlphabet()), id='chrGL000239.1', name='chrGL000239.1', description='chrGL000239.1 dna:supercontig supercontig::GL000239.1:1:33824:1', dbxrefs=[])),\n",
       "             ('chrGL000235.1',\n",
       "              SeqRecord(seq=Seq('GATCTCTGAGGTGATGCAACTCTTGTCTAGGCACTGCCTACAGGGGACATTGGT...ATC', SingleLetterAlphabet()), id='chrGL000235.1', name='chrGL000235.1', description='chrGL000235.1 dna:supercontig supercontig::GL000235.1:1:34474:1', dbxrefs=[])),\n",
       "             ('chrGL000201.1',\n",
       "              SeqRecord(seq=Seq('GATCCCACCCAAAGCAGCAGCTTTTTTTTTTTTTTTTTTTTTTTTTGTTATTTT...ATC', SingleLetterAlphabet()), id='chrGL000201.1', name='chrGL000201.1', description='chrGL000201.1 dna:supercontig supercontig::GL000201.1:1:36148:1', dbxrefs=[])),\n",
       "             ('chrGL000247.1',\n",
       "              SeqRecord(seq=Seq('GATCTAAGTTTGATTTTTCTGTACAACTCCTTTTCTGAAGTTTTACTCATTTCC...ATC', SingleLetterAlphabet()), id='chrGL000247.1', name='chrGL000247.1', description='chrGL000247.1 dna:supercontig supercontig::GL000247.1:1:36422:1', dbxrefs=[])),\n",
       "             ('chrGL000245.1',\n",
       "              SeqRecord(seq=Seq('GATCAGATAGGCTTTTGGCTCATTGTAATGAGATGTTCCTATAAACCTAATTTG...ATC', SingleLetterAlphabet()), id='chrGL000245.1', name='chrGL000245.1', description='chrGL000245.1 dna:supercontig supercontig::GL000245.1:1:36651:1', dbxrefs=[])),\n",
       "             ('chrGL000197.1',\n",
       "              SeqRecord(seq=Seq('CAAGACTCCTCAGAGTAAGGCATCTGTCCTATTTTGTTCCCCACCCAGGAGTTG...TGG', SingleLetterAlphabet()), id='chrGL000197.1', name='chrGL000197.1', description='chrGL000197.1 dna:supercontig supercontig::GL000197.1:1:37175:1', dbxrefs=[])),\n",
       "             ('chrGL000203.1',\n",
       "              SeqRecord(seq=Seq('GATCCCACAGGATGTGTCTATCTTAAGATTTAGATAAATATTTAAATTTTTTTC...ATC', SingleLetterAlphabet()), id='chrGL000203.1', name='chrGL000203.1', description='chrGL000203.1 dna:supercontig supercontig::GL000203.1:1:37498:1', dbxrefs=[])),\n",
       "             ('chrGL000246.1',\n",
       "              SeqRecord(seq=Seq('GATCTTAAGCCTTTGCTGGATGCATCGTTTGCAAATATCTTCTCCCATTGTGTA...ATC', SingleLetterAlphabet()), id='chrGL000246.1', name='chrGL000246.1', description='chrGL000246.1 dna:supercontig supercontig::GL000246.1:1:38154:1', dbxrefs=[])),\n",
       "             ('chrGL000249.1',\n",
       "              SeqRecord(seq=Seq('GATCACCAAGGCTGGGAAATATTAGGGAAGGGGGGTTGCCAGACTCCGGACCTG...ATC', SingleLetterAlphabet()), id='chrGL000249.1', name='chrGL000249.1', description='chrGL000249.1 dna:supercontig supercontig::GL000249.1:1:38502:1', dbxrefs=[])),\n",
       "             ('chrGL000196.1',\n",
       "              SeqRecord(seq=Seq('GATCCCACCAGAGTAGAGGTGGCTATACTAACAACAGAGAAAACAGTCTAAATC...ATC', SingleLetterAlphabet()), id='chrGL000196.1', name='chrGL000196.1', description='chrGL000196.1 dna:supercontig supercontig::GL000196.1:1:38914:1', dbxrefs=[])),\n",
       "             ('chrGL000248.1',\n",
       "              SeqRecord(seq=Seq('GATCTGTCATTGTCTCCCATGCATGCTTCCCAGCATGCCTGCCTAGCCCCATTC...ATC', SingleLetterAlphabet()), id='chrGL000248.1', name='chrGL000248.1', description='chrGL000248.1 dna:supercontig supercontig::GL000248.1:1:39786:1', dbxrefs=[])),\n",
       "             ('chrGL000244.1',\n",
       "              SeqRecord(seq=Seq('GATCTGCGCACCTTGGCCTCCCAAAGTGCTGGGATTACAGGCATGAGCCACCGT...ATC', SingleLetterAlphabet()), id='chrGL000244.1', name='chrGL000244.1', description='chrGL000244.1 dna:supercontig supercontig::GL000244.1:1:39929:1', dbxrefs=[])),\n",
       "             ('chrGL000238.1',\n",
       "              SeqRecord(seq=Seq('GATCACAGTATCACAATTGACCTTAAAACAATAAATTTTATAATCAGTGTAAAT...ATC', SingleLetterAlphabet()), id='chrGL000238.1', name='chrGL000238.1', description='chrGL000238.1 dna:supercontig supercontig::GL000238.1:1:39939:1', dbxrefs=[])),\n",
       "             ('chrGL000202.1',\n",
       "              SeqRecord(seq=Seq('TGCCAGAAAAGGGTGTGTGTTGGAGGGGGTGTCTTTCACAATATATTTTTAATT...AGC', SingleLetterAlphabet()), id='chrGL000202.1', name='chrGL000202.1', description='chrGL000202.1 dna:supercontig supercontig::GL000202.1:1:40103:1', dbxrefs=[])),\n",
       "             ('chrGL000234.1',\n",
       "              SeqRecord(seq=Seq('GATCTCAATGCATCCAGATTCATAATATGTTTGCCTTTGAATGGAATGATCCAG...ATC', SingleLetterAlphabet()), id='chrGL000234.1', name='chrGL000234.1', description='chrGL000234.1 dna:supercontig supercontig::GL000234.1:1:40531:1', dbxrefs=[])),\n",
       "             ('chrGL000232.1',\n",
       "              SeqRecord(seq=Seq('GATCAGGCCCAGGAACTGCTGCCTTGGGTAGCTTTGGAACAAGGCTCCTGCATC...ATC', SingleLetterAlphabet()), id='chrGL000232.1', name='chrGL000232.1', description='chrGL000232.1 dna:supercontig supercontig::GL000232.1:1:40652:1', dbxrefs=[])),\n",
       "             ('chrGL000206.1',\n",
       "              SeqRecord(seq=Seq('GATCCCACCAGAGGCCATAAACAGATAGTTCAGAGCTATGAGAACCCAGGTGGC...ATC', SingleLetterAlphabet()), id='chrGL000206.1', name='chrGL000206.1', description='chrGL000206.1 dna:supercontig supercontig::GL000206.1:1:41001:1', dbxrefs=[])),\n",
       "             ('chrGL000240.1',\n",
       "              SeqRecord(seq=Seq('GATCAGTGGAGACCTGGTCAGCTGGGGCTTAGTGCTGGCCTGGTCAGCATGGGC...ATC', SingleLetterAlphabet()), id='chrGL000240.1', name='chrGL000240.1', description='chrGL000240.1 dna:supercontig supercontig::GL000240.1:1:41933:1', dbxrefs=[])),\n",
       "             ('chrGL000236.1',\n",
       "              SeqRecord(seq=Seq('GATCACATGGACACAGGAAGGGGAATATCACACTCTGGGGACTGTTGTGGGGTG...ATC', SingleLetterAlphabet()), id='chrGL000236.1', name='chrGL000236.1', description='chrGL000236.1 dna:supercontig supercontig::GL000236.1:1:41934:1', dbxrefs=[])),\n",
       "             ('chrGL000241.1',\n",
       "              SeqRecord(seq=Seq('GATCTGAAGACAAATCAGGGTTGCAGGATGGTTAAATATCTGCAAGTCAATAAA...ATC', SingleLetterAlphabet()), id='chrGL000241.1', name='chrGL000241.1', description='chrGL000241.1 dna:supercontig supercontig::GL000241.1:1:42152:1', dbxrefs=[])),\n",
       "             ('chrGL000243.1',\n",
       "              SeqRecord(seq=Seq('GATCTTCAGGCTCTGTCTCTGATTCTAGTGCTCTTGTTATTTCCACCATATCTG...ATC', SingleLetterAlphabet()), id='chrGL000243.1', name='chrGL000243.1', description='chrGL000243.1 dna:supercontig supercontig::GL000243.1:1:43341:1', dbxrefs=[])),\n",
       "             ('chrGL000242.1',\n",
       "              SeqRecord(seq=Seq('GATCATGCCTGGGACACACAGCAAGACCTTGTCCCTTAAAAAAAAAAAAAAAAA...ATC', SingleLetterAlphabet()), id='chrGL000242.1', name='chrGL000242.1', description='chrGL000242.1 dna:supercontig supercontig::GL000242.1:1:43523:1', dbxrefs=[])),\n",
       "             ('chrGL000230.1',\n",
       "              SeqRecord(seq=Seq('GATCCTTTGAGCCCAGGAGTTTGAGACCACCCTGGGAAACATGGCAAAACCCTC...ATC', SingleLetterAlphabet()), id='chrGL000230.1', name='chrGL000230.1', description='chrGL000230.1 dna:supercontig supercontig::GL000230.1:1:43691:1', dbxrefs=[])),\n",
       "             ('chrGL000237.1',\n",
       "              SeqRecord(seq=Seq('GATCTCAGGAGGCGGAGGTTGCAGTGATCGGTGGCCATAGAAGCCCTCCTGAGC...ATC', SingleLetterAlphabet()), id='chrGL000237.1', name='chrGL000237.1', description='chrGL000237.1 dna:supercontig supercontig::GL000237.1:1:45867:1', dbxrefs=[])),\n",
       "             ('chrGL000233.1',\n",
       "              SeqRecord(seq=Seq('GATCCCATTTGTCAATTTTGGATTTTGTGGCCATTGCTTCTGTTGTTTTAGACA...ATC', SingleLetterAlphabet()), id='chrGL000233.1', name='chrGL000233.1', description='chrGL000233.1 dna:supercontig supercontig::GL000233.1:1:45941:1', dbxrefs=[])),\n",
       "             ('chrGL000204.1',\n",
       "              SeqRecord(seq=Seq('GATCAGGAAGAGTCCTGGGGGGCGGGGGGCGAGCACAGGCAGTGGAAGGGAAGC...ATC', SingleLetterAlphabet()), id='chrGL000204.1', name='chrGL000204.1', description='chrGL000204.1 dna:supercontig supercontig::GL000204.1:1:81310:1', dbxrefs=[])),\n",
       "             ('chrGL000198.1',\n",
       "              SeqRecord(seq=Seq('GAATTCGTCTCATAGAGTTGAACCTTTCTTTTCATTGATCAGTTTGGAAACACT...TTC', SingleLetterAlphabet()), id='chrGL000198.1', name='chrGL000198.1', description='chrGL000198.1 dna:supercontig supercontig::GL000198.1:1:90085:1', dbxrefs=[])),\n",
       "             ('chrGL000208.1',\n",
       "              SeqRecord(seq=Seq('GAATTCTTCAAAGAGTTCCAGATATCCACAGGCAGATTCTACAAAAGAAGTGTT...GAT', SingleLetterAlphabet()), id='chrGL000208.1', name='chrGL000208.1', description='chrGL000208.1 dna:supercontig supercontig::GL000208.1:1:92689:1', dbxrefs=[])),\n",
       "             ('chrGL000191.1',\n",
       "              SeqRecord(seq=Seq('GATCCACCTGCCTCAGCCTCCCAGAGTGCTGGGATTATAGGTGTGAGCCACTGC...ATC', SingleLetterAlphabet()), id='chrGL000191.1', name='chrGL000191.1', description='chrGL000191.1 dna:supercontig supercontig::GL000191.1:1:106433:1', dbxrefs=[])),\n",
       "             ('chrGL000227.1',\n",
       "              SeqRecord(seq=Seq('GATCCAACTATATTGTCCACAAGATACTCACTTTAGATCAAAATACACAATGAG...ATC', SingleLetterAlphabet()), id='chrGL000227.1', name='chrGL000227.1', description='chrGL000227.1 dna:supercontig supercontig::GL000227.1:1:128374:1', dbxrefs=[])),\n",
       "             ('chrGL000228.1',\n",
       "              SeqRecord(seq=Seq('GAATTCTGTGCATTTTGTTTGATGGAATTTGGTATGTTTACCAAAATGGCAGAA...TTC', SingleLetterAlphabet()), id='chrGL000228.1', name='chrGL000228.1', description='chrGL000228.1 dna:supercontig supercontig::GL000228.1:1:129120:1', dbxrefs=[])),\n",
       "             ('chrGL000214.1',\n",
       "              SeqRecord(seq=Seq('GAATTCAGTGCAATTCAAAACAGTAATTTGTGGGAAAAATTAGATATGCAGATA...TTC', SingleLetterAlphabet()), id='chrGL000214.1', name='chrGL000214.1', description='chrGL000214.1 dna:supercontig supercontig::GL000214.1:1:137718:1', dbxrefs=[])),\n",
       "             ('chrGL000221.1',\n",
       "              SeqRecord(seq=Seq('GAATTCGTTGAGGAGCTTCTGGAAAGTGCACATTCTGACTCAGCAGGTATTGGA...TTC', SingleLetterAlphabet()), id='chrGL000221.1', name='chrGL000221.1', description='chrGL000221.1 dna:supercontig supercontig::GL000221.1:1:155397:1', dbxrefs=[])),\n",
       "             ('chrGL000209.1',\n",
       "              SeqRecord(seq=Seq('GATCAGCTTTCCCTTTGAAATCAGAGTACTAATCTGACTCATTGTTCCCTGAAT...ATC', SingleLetterAlphabet()), id='chrGL000209.1', name='chrGL000209.1', description='chrGL000209.1 dna:supercontig supercontig::GL000209.1:1:159169:1', dbxrefs=[])),\n",
       "             ('chrGL000218.1',\n",
       "              SeqRecord(seq=Seq('GAATTCATGTAAAGAAAGACCATGGGTGTTTGTACTGGATTTAGTATTCATCAT...TTC', SingleLetterAlphabet()), id='chrGL000218.1', name='chrGL000218.1', description='chrGL000218.1 dna:supercontig supercontig::GL000218.1:1:161147:1', dbxrefs=[])),\n",
       "             ('chrGL000220.1',\n",
       "              SeqRecord(seq=Seq('GAATTCAATACTTAAACATAAAACCTAATAAAATTCTTAGAAGTAAACATAAGG...TTC', SingleLetterAlphabet()), id='chrGL000220.1', name='chrGL000220.1', description='chrGL000220.1 dna:supercontig supercontig::GL000220.1:1:161802:1', dbxrefs=[])),\n",
       "             ('chrGL000213.1',\n",
       "              SeqRecord(seq=Seq('GAATTCAGTGAATGAGCCAAGATTCTCCAGAAACCGCTCAGCTTTGCAGTGATA...TTC', SingleLetterAlphabet()), id='chrGL000213.1', name='chrGL000213.1', description='chrGL000213.1 dna:supercontig supercontig::GL000213.1:1:164239:1', dbxrefs=[])),\n",
       "             ('chrGL000211.1',\n",
       "              SeqRecord(seq=Seq('GAATTCTTTTTCAAGTAAGTCAGGGATTTCTTCTTAGTTTGGATCCATTGCTGG...TTC', SingleLetterAlphabet()), id='chrGL000211.1', name='chrGL000211.1', description='chrGL000211.1 dna:supercontig supercontig::GL000211.1:1:166566:1', dbxrefs=[])),\n",
       "             ('chrGL000199.1',\n",
       "              SeqRecord(seq=Seq('GAATTCTATATGAGGGAGAAAAACTCAGAACCTAGCAGCAGTGTTCTGGAATCC...TTC', SingleLetterAlphabet()), id='chrGL000199.1', name='chrGL000199.1', description='chrGL000199.1 dna:supercontig supercontig::GL000199.1:1:169874:1', dbxrefs=[])),\n",
       "             ('chrGL000217.1',\n",
       "              SeqRecord(seq=Seq('GAATTCACTGGTCTTACCATATTCCCCATCTTCCTGAAGCAGCTGGATTGATAG...TTC', SingleLetterAlphabet()), id='chrGL000217.1', name='chrGL000217.1', description='chrGL000217.1 dna:supercontig supercontig::GL000217.1:1:172149:1', dbxrefs=[])),\n",
       "             ('chrGL000216.1',\n",
       "              SeqRecord(seq=Seq('GAATTCCATTGCATTCCATTCCATTTCATTCTATTACACTCCATTCCATTCCTT...TTC', SingleLetterAlphabet()), id='chrGL000216.1', name='chrGL000216.1', description='chrGL000216.1 dna:supercontig supercontig::GL000216.1:1:172294:1', dbxrefs=[])),\n",
       "             ('chrGL000215.1',\n",
       "              SeqRecord(seq=Seq('GAATTCCATAATATAAAGTTACTTTTTTAAAAAAAAATTAGACAATCAATTATC...TTC', SingleLetterAlphabet()), id='chrGL000215.1', name='chrGL000215.1', description='chrGL000215.1 dna:supercontig supercontig::GL000215.1:1:172545:1', dbxrefs=[])),\n",
       "             ('chrGL000205.1',\n",
       "              SeqRecord(seq=Seq('GAATTCTTGGTATTATTCCTTTTGAGTACCCCTATTTATGAGAGAGAAAACGAA...TTC', SingleLetterAlphabet()), id='chrGL000205.1', name='chrGL000205.1', description='chrGL000205.1 dna:supercontig supercontig::GL000205.1:1:174588:1', dbxrefs=[])),\n",
       "             ('chrGL000219.1',\n",
       "              SeqRecord(seq=Seq('GAATTCTACCTGTGGGACAAACATTGAGACCTTTGTAACATTGTTCTGGAATGC...TTC', SingleLetterAlphabet()), id='chrGL000219.1', name='chrGL000219.1', description='chrGL000219.1 dna:supercontig supercontig::GL000219.1:1:179198:1', dbxrefs=[])),\n",
       "             ('chrGL000224.1',\n",
       "              SeqRecord(seq=Seq('GAATTCTGTTGATGATGGATGTGAGAGTGGTGAAGCCCTGCGGGATGATGTAGA...TTC', SingleLetterAlphabet()), id='chrGL000224.1', name='chrGL000224.1', description='chrGL000224.1 dna:supercontig supercontig::GL000224.1:1:179693:1', dbxrefs=[])),\n",
       "             ('chrGL000223.1',\n",
       "              SeqRecord(seq=Seq('GAATTCTCTGATGTCGAGTGAGGTTGGAGGCACGGCTAAAGGCCTTTCCACACT...TTC', SingleLetterAlphabet()), id='chrGL000223.1', name='chrGL000223.1', description='chrGL000223.1 dna:supercontig supercontig::GL000223.1:1:180455:1', dbxrefs=[])),\n",
       "             ('chrGL000195.1',\n",
       "              SeqRecord(seq=Seq('GAATTCCTCGTTCACACAGTTTCTTAAGCTTCCTGGGATGCGACCTGTGATGGC...TTC', SingleLetterAlphabet()), id='chrGL000195.1', name='chrGL000195.1', description='chrGL000195.1 dna:supercontig supercontig::GL000195.1:1:182896:1', dbxrefs=[])),\n",
       "             ('chrGL000212.1',\n",
       "              SeqRecord(seq=Seq('GAATTCCTCCTTGCACATTTCTGCTTTATCTCATCTACACAGAAGTGATCCAAT...TTC', SingleLetterAlphabet()), id='chrGL000212.1', name='chrGL000212.1', description='chrGL000212.1 dna:supercontig supercontig::GL000212.1:1:186858:1', dbxrefs=[])),\n",
       "             ('chrGL000222.1',\n",
       "              SeqRecord(seq=Seq('GAATTCTATGTGAGGGACACACATTCAAAACCCAGCAGCAGTATTCTGGAATCC...TTC', SingleLetterAlphabet()), id='chrGL000222.1', name='chrGL000222.1', description='chrGL000222.1 dna:supercontig supercontig::GL000222.1:1:186861:1', dbxrefs=[])),\n",
       "             ('chrGL000200.1',\n",
       "              SeqRecord(seq=Seq('TGGGTTGATAGGTTTTTAAGAATTCTTTTATGACTTCAAAAGTGTCACTTCGTT...CCT', SingleLetterAlphabet()), id='chrGL000200.1', name='chrGL000200.1', description='chrGL000200.1 dna:supercontig supercontig::GL000200.1:1:187035:1', dbxrefs=[])),\n",
       "             ('chrGL000193.1',\n",
       "              SeqRecord(seq=Seq('GAATTCATTTCTTTGATTTGTAGAGTGGCAGGGTACCTAAAGTTAACATTTGTC...TTC', SingleLetterAlphabet()), id='chrGL000193.1', name='chrGL000193.1', description='chrGL000193.1 dna:supercontig supercontig::GL000193.1:1:189789:1', dbxrefs=[])),\n",
       "             ('chrGL000194.1',\n",
       "              SeqRecord(seq=Seq('GATCCCTGCCCTAAAACTTTCCCCCCTCATGTCCAGCAAATGCTGCATGGAGCC...ATC', SingleLetterAlphabet()), id='chrGL000194.1', name='chrGL000194.1', description='chrGL000194.1 dna:supercontig supercontig::GL000194.1:1:191469:1', dbxrefs=[])),\n",
       "             ('chrGL000225.1',\n",
       "              SeqRecord(seq=Seq('GAATTCTCACATGTCTCTGCACTGATCACCCAGGTGATGTAACTCCTGTCTAGG...TTC', SingleLetterAlphabet()), id='chrGL000225.1', name='chrGL000225.1', description='chrGL000225.1 dna:supercontig supercontig::GL000225.1:1:211173:1', dbxrefs=[])),\n",
       "             ('chrGL000192.1',\n",
       "              SeqRecord(seq=Seq('GAATTCATTCACCATTATTCTTTTATAATATTGCTATTTTATTATTCTTGATCA...TTC', SingleLetterAlphabet()), id='chrGL000192.1', name='chrGL000192.1', description='chrGL000192.1 dna:supercontig supercontig::GL000192.1:1:547496:1', dbxrefs=[]))])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomeDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sublibrary</th>\n",
       "      <th>gene</th>\n",
       "      <th>transcripts</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sgId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323216.24-all~e39m1</th>\n",
       "      <td>drug_targets+kinase_phosphatase</td>\n",
       "      <td>AARS</td>\n",
       "      <td>all</td>\n",
       "      <td>GCCCCAGGATCAGGCCCCGCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323296.24-all~e39m1</th>\n",
       "      <td>drug_targets+kinase_phosphatase</td>\n",
       "      <td>AARS</td>\n",
       "      <td>all</td>\n",
       "      <td>GGCCGCCCTCGGAGAGCTCTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323318.24-all~e39m1</th>\n",
       "      <td>drug_targets+kinase_phosphatase</td>\n",
       "      <td>AARS</td>\n",
       "      <td>all</td>\n",
       "      <td>GACGGCGACCCTAGGAGAGGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323362.24-all~e39m1</th>\n",
       "      <td>drug_targets+kinase_phosphatase</td>\n",
       "      <td>AARS</td>\n",
       "      <td>all</td>\n",
       "      <td>GGTGCAGCGGGCCCTTGGCGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323441.24-all~e39m1</th>\n",
       "      <td>drug_targets+kinase_phosphatase</td>\n",
       "      <td>AARS</td>\n",
       "      <td>all</td>\n",
       "      <td>GCGCTCTGATTGGACGGAGCG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                         sublibrary  \\\n",
       "sgId                                                                                  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...  drug_targets+kinase_phosphatase   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...  drug_targets+kinase_phosphatase   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...  drug_targets+kinase_phosphatase   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...  drug_targets+kinase_phosphatase   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...  drug_targets+kinase_phosphatase   \n",
       "\n",
       "                                                    gene transcripts  \\\n",
       "sgId                                                                   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...  AARS         all   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...  AARS         all   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...  AARS         all   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...  AARS         all   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...  AARS         all   \n",
       "\n",
       "                                                                 sequence  \n",
       "sgId                                                                       \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...  GCCCCAGGATCAGGCCCCGCG  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...  GGCCGCCCTCGGAGAGCTCTG  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...  GACGGCGACCCTAGGAGAGGT  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...  GGTGCAGCGGGCCCTTGGCGG  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...  GCGCTCTGATTGGACGGAGCG  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#to use pre-calculated sgRNA activity score data (e.g. provided CRISPRi training data), load the following:\n",
    "#CRISPRa activity score data also included in data_files\n",
    "libraryTable_training = pd.read_csv('data_files/CRISPRi_trainingdata_libraryTable.txt', sep='\\t', index_col = 0)\n",
    "libraryTable_training.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sublibrary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>length</th>\n",
       "      <th>pam coordinate</th>\n",
       "      <th>pass_score</th>\n",
       "      <th>position</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_list</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sgId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323216.24-all~e39m1</th>\n",
       "      <td>Drug_Targets+Kinase_Phosphatase</td>\n",
       "      <td>AARS</td>\n",
       "      <td>24</td>\n",
       "      <td>70323216</td>\n",
       "      <td>e39m1</td>\n",
       "      <td>70323216</td>\n",
       "      <td>+</td>\n",
       "      <td>['all']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323296.24-all~e39m1</th>\n",
       "      <td>Drug_Targets+Kinase_Phosphatase</td>\n",
       "      <td>AARS</td>\n",
       "      <td>24</td>\n",
       "      <td>70323296</td>\n",
       "      <td>e39m1</td>\n",
       "      <td>70323296</td>\n",
       "      <td>+</td>\n",
       "      <td>['all']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323318.24-all~e39m1</th>\n",
       "      <td>Drug_Targets+Kinase_Phosphatase</td>\n",
       "      <td>AARS</td>\n",
       "      <td>24</td>\n",
       "      <td>70323318</td>\n",
       "      <td>e39m1</td>\n",
       "      <td>70323318</td>\n",
       "      <td>+</td>\n",
       "      <td>['all']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323362.24-all~e39m1</th>\n",
       "      <td>Drug_Targets+Kinase_Phosphatase</td>\n",
       "      <td>AARS</td>\n",
       "      <td>24</td>\n",
       "      <td>70323362</td>\n",
       "      <td>e39m1</td>\n",
       "      <td>70323362</td>\n",
       "      <td>+</td>\n",
       "      <td>['all']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323441.24-all~e39m1</th>\n",
       "      <td>Drug_Targets+Kinase_Phosphatase</td>\n",
       "      <td>AARS</td>\n",
       "      <td>24</td>\n",
       "      <td>70323441</td>\n",
       "      <td>e39m1</td>\n",
       "      <td>70323441</td>\n",
       "      <td>+</td>\n",
       "      <td>['all']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                         Sublibrary  \\\n",
       "sgId                                                                                  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...  Drug_Targets+Kinase_Phosphatase   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...  Drug_Targets+Kinase_Phosphatase   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...  Drug_Targets+Kinase_Phosphatase   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...  Drug_Targets+Kinase_Phosphatase   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...  Drug_Targets+Kinase_Phosphatase   \n",
       "\n",
       "                                                   gene_name  length  \\\n",
       "sgId                                                                   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...      AARS      24   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...      AARS      24   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...      AARS      24   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...      AARS      24   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...      AARS      24   \n",
       "\n",
       "                                                    pam coordinate pass_score  \\\n",
       "sgId                                                                            \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...        70323216      e39m1   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...        70323296      e39m1   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...        70323318      e39m1   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...        70323362      e39m1   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...        70323441      e39m1   \n",
       "\n",
       "                                                    position strand  \\\n",
       "sgId                                                                  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...  70323216      +   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...  70323296      +   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...  70323318      +   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...  70323362      +   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...  70323441      +   \n",
       "\n",
       "                                                   transcript_list  \n",
       "sgId                                                                \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...         ['all']  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...         ['all']  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...         ['all']  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...         ['all']  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...         ['all']  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sgInfoTable_training = pd.read_csv('data_files/CRISPRi_trainingdata_sgRNAInfoTable.txt', sep='\\t', index_col=0)\n",
    "sgInfoTable_training.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216.24-all~e39m1    0.348892\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296.24-all~e39m1    0.912409\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318.24-all~e39m1    0.997242\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362.24-all~e39m1    0.962154\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441.24-all~e39m1    0.019320\n",
       "Name: 1, dtype: float64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "activityScores = pd.read_csv('data_files/CRISPRi_trainingdata_activityScores.txt',sep='\\t',index_col=0, header=None).iloc[:,0]\n",
    "activityScores.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>position</th>\n",
       "      <th>strand</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>cage peak ranges</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th>transcripts</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <th>all</th>\n",
       "      <td>58864864.0</td>\n",
       "      <td>-</td>\n",
       "      <td>chr19</td>\n",
       "      <td>[(58864822, 58864847), (58864848, 58864868)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">A1CF</th>\n",
       "      <th>ENST00000373993.1</th>\n",
       "      <td>52619744.0</td>\n",
       "      <td>-</td>\n",
       "      <td>chr10</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000374001.2,ENST00000373997.3,ENST00000282641.2</th>\n",
       "      <td>52645434.0</td>\n",
       "      <td>-</td>\n",
       "      <td>chr10</td>\n",
       "      <td>[(52645379, 52645393), (52645416, 52645444)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <th>all</th>\n",
       "      <td>9268752.0</td>\n",
       "      <td>-</td>\n",
       "      <td>chr12</td>\n",
       "      <td>[(9268547, 9268556), (9268559, 9268568), (9268...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <th>all</th>\n",
       "      <td>8975067.0</td>\n",
       "      <td>+</td>\n",
       "      <td>chr12</td>\n",
       "      <td>[(8975061, 8975072), (8975101, 8975108), (8975...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                            position strand  \\\n",
       "gene  transcripts                                                             \n",
       "A1BG  all                                                 58864864.0      -   \n",
       "A1CF  ENST00000373993.1                                   52619744.0      -   \n",
       "      ENST00000374001.2,ENST00000373997.3,ENST0000028...  52645434.0      -   \n",
       "A2M   all                                                  9268752.0      -   \n",
       "A2ML1 all                                                  8975067.0      +   \n",
       "\n",
       "                                                         chromosome  \\\n",
       "gene  transcripts                                                     \n",
       "A1BG  all                                                     chr19   \n",
       "A1CF  ENST00000373993.1                                       chr10   \n",
       "      ENST00000374001.2,ENST00000373997.3,ENST0000028...      chr10   \n",
       "A2M   all                                                     chr12   \n",
       "A2ML1 all                                                     chr12   \n",
       "\n",
       "                                                                                           cage peak ranges  \n",
       "gene  transcripts                                                                                            \n",
       "A1BG  all                                                      [(58864822, 58864847), (58864848, 58864868)]  \n",
       "A1CF  ENST00000373993.1                                                                                  []  \n",
       "      ENST00000374001.2,ENST00000373997.3,ENST0000028...       [(52645379, 52645393), (52645416, 52645444)]  \n",
       "A2M   all                                                 [(9268547, 9268556), (9268559, 9268568), (9268...  \n",
       "A2ML1 all                                                 [(8975061, 8975072), (8975101, 8975108), (8975...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tssTable = pd.read_csv('data_files/human_tssTable.txt',sep='\\t', index_col=range(2))\n",
    "tssTable.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>strand</th>\n",
       "      <th>TSS source</th>\n",
       "      <th>primary TSS</th>\n",
       "      <th>secondary TSS</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th>transcript</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">A1BG</th>\n",
       "      <th>P1</th>\n",
       "      <td>chr19</td>\n",
       "      <td>-</td>\n",
       "      <td>CAGE, matched peaks</td>\n",
       "      <td>(58858938, 58859039)</td>\n",
       "      <td>(58858938, 58859039)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2</th>\n",
       "      <td>chr19</td>\n",
       "      <td>-</td>\n",
       "      <td>CAGE, matched peaks</td>\n",
       "      <td>(58864822, 58864847)</td>\n",
       "      <td>(58864822, 58864847)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <th>P1P2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>-</td>\n",
       "      <td>CAGE, matched peaks</td>\n",
       "      <td>(52645379, 52645393)</td>\n",
       "      <td>(52645379, 52645393)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <th>P1P2</th>\n",
       "      <td>chr12</td>\n",
       "      <td>-</td>\n",
       "      <td>CAGE, matched peaks</td>\n",
       "      <td>(9268507, 9268523)</td>\n",
       "      <td>(9268528, 9268542)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <th>P1P2</th>\n",
       "      <td>chr12</td>\n",
       "      <td>+</td>\n",
       "      <td>CAGE, matched peaks</td>\n",
       "      <td>(8975206, 8975223)</td>\n",
       "      <td>(8975144, 8975169)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 chromosome strand           TSS source           primary TSS  \\\n",
       "gene  transcript                                                                \n",
       "A1BG  P1              chr19      -  CAGE, matched peaks  (58858938, 58859039)   \n",
       "      P2              chr19      -  CAGE, matched peaks  (58864822, 58864847)   \n",
       "A1CF  P1P2            chr10      -  CAGE, matched peaks  (52645379, 52645393)   \n",
       "A2M   P1P2            chr12      -  CAGE, matched peaks    (9268507, 9268523)   \n",
       "A2ML1 P1P2            chr12      +  CAGE, matched peaks    (8975206, 8975223)   \n",
       "\n",
       "                         secondary TSS  \n",
       "gene  transcript                        \n",
       "A1BG  P1          (58858938, 58859039)  \n",
       "      P2          (58864822, 58864847)  \n",
       "A1CF  P1P2        (52645379, 52645393)  \n",
       "A2M   P1P2          (9268528, 9268542)  \n",
       "A2ML1 P1P2          (8975144, 8975169)  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p1p2Table = pd.read_csv('data_files/human_p1p2Table.txt',sep='\\t', header=0, index_col=range(2))\n",
    "p1p2Table['primary TSS'] = p1p2Table['primary TSS'].apply(lambda tupString: (int(tupString.strip('()').split(', ')[0].split('.')[0]), int(tupString.strip('()').split(', ')[1].split('.')[0])))\n",
    "p1p2Table['secondary TSS'] = p1p2Table['secondary TSS'].apply(lambda tupString: (int(tupString.strip('()').split(', ')[0].split('.')[0]),int(tupString.strip('()').split(', ')[1].split('.')[0])))\n",
    "p1p2Table.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Calculate parameters for empirical sgRNAs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Because scikit-learn currently does not support any robust method for saving and re-loading the machine learning model, the best strategy is to simply re-learn the model from the training data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above statement is untrue, pickle is often used to same models."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load bigwig files for any chromatin data of interest\n",
    "bwhandleDict = {'dnase':BigWigFile(open('large_data_files/wgEncodeOpenChromDnaseK562BaseOverlapSignalV2.bigWig')),\n",
    "'faire':BigWigFile(open('large_data_files/wgEncodeOpenChromFaireK562Sig.bigWig')),\n",
    "'mnase':BigWigFile(open('large_data_files/wgEncodeSydhNsomeK562Sig.bigWig'))}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run generateTypicalParamTable manually"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "libraryTable = libraryTable_training\n",
    "sgInfoTable = sgInfoTable_training\n",
    "bwFileHandleDict = bwhandleDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "lengthSeries = generateSgrnaLengthSeries(libraryTable)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sgId\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216.24-all~e39m1    21\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296.24-all~e39m1    21\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318.24-all~e39m1    21\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362.24-all~e39m1    21\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441.24-all~e39m1    21\n",
       "Name: length, dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lengthSeries.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "sgrnaPositionTable_p1p2 = generateSgrnaDistanceTable_p1p2Strategy(sgInfoTable,\n",
    "                                                                  libraryTable,\n",
    "                                                                  p1p2Table,\n",
    "                                                                  False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript</th>\n",
       "      <th>primary TSS-Up</th>\n",
       "      <th>primary TSS-Down</th>\n",
       "      <th>secondary TSS-Up</th>\n",
       "      <th>secondary TSS-Down</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sgId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323216.24-all~e39m1</th>\n",
       "      <td>AARS</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>205</td>\n",
       "      <td>167</td>\n",
       "      <td>247</td>\n",
       "      <td>206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323296.24-all~e39m1</th>\n",
       "      <td>AARS</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>125</td>\n",
       "      <td>87</td>\n",
       "      <td>167</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323318.24-all~e39m1</th>\n",
       "      <td>AARS</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>103</td>\n",
       "      <td>65</td>\n",
       "      <td>145</td>\n",
       "      <td>104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323362.24-all~e39m1</th>\n",
       "      <td>AARS</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>59</td>\n",
       "      <td>21</td>\n",
       "      <td>101</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323441.24-all~e39m1</th>\n",
       "      <td>AARS</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>-20</td>\n",
       "      <td>-58</td>\n",
       "      <td>22</td>\n",
       "      <td>-19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    gene transcript  \\\n",
       "sgId                                                                  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...  AARS       P1P2   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...  AARS       P1P2   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...  AARS       P1P2   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...  AARS       P1P2   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...  AARS       P1P2   \n",
       "\n",
       "                                                    primary TSS-Up  \\\n",
       "sgId                                                                 \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...             205   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...             125   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...             103   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...              59   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...             -20   \n",
       "\n",
       "                                                    primary TSS-Down  \\\n",
       "sgId                                                                   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...               167   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...                87   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...                65   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...                21   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...               -58   \n",
       "\n",
       "                                                    secondary TSS-Up  \\\n",
       "sgId                                                                   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...               247   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...               167   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...               145   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...               101   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...                22   \n",
       "\n",
       "                                                    secondary TSS-Down  \n",
       "sgId                                                                    \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...                 206  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...                 126  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...                 104  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...                  60  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...                 -19  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sgrnaPositionTable_p1p2.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "baseTable, strand = generateRelativeBasesAndStrand(sgInfoTable,\n",
    "                                                   tssTable,\n",
    "                                                   libraryTable, genomeDict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>-30</th>\n",
       "      <th>-29</th>\n",
       "      <th>-28</th>\n",
       "      <th>-27</th>\n",
       "      <th>-26</th>\n",
       "      <th>-25</th>\n",
       "      <th>-24</th>\n",
       "      <th>-23</th>\n",
       "      <th>-22</th>\n",
       "      <th>-21</th>\n",
       "      <th>...</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sgId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323216.24-all~e39m1</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323296.24-all~e39m1</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323318.24-all~e39m1</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323362.24-all~e39m1</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323441.24-all~e39m1</th>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   -30 -29 -28 -27 -26 -25  \\\n",
       "sgId                                                                         \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...   G   C   A   G   T   G   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...   A   G   A   G   G   T   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...   A   G   C   T   G   G   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...   G   G   G   A   A   T   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...   C   G   T   C   G   G   \n",
       "\n",
       "                                                   -24 -23 -22 -21  ...  0   \\\n",
       "sgId                                                                ...       \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...   G   G   C   C  ...   G   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...   A   G   G   C  ...   G   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...   G   G   A   C  ...   G   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...   A   G   G   T  ...   G   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...   C   G   C   G  ...   G   \n",
       "\n",
       "                                                    1   2   3   4   5   6   \\\n",
       "sgId                                                                         \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...   C   C   T   G   G   A   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...   T   G   C   T   A   G   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...   C   C   G   C   C   C   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...   A   C   T   C   T   G   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...   C   G   G   G   T   T   \n",
       "\n",
       "                                                    7   8   9   \n",
       "sgId                                                            \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...   G   A   A  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...   G   A   G  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...   T   C   G  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...   A   G   G  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...   G   T   G  \n",
       "\n",
       "[5 rows x 40 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "baseTable.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>same strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sgId</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323216.24-all~e39m1</th>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323296.24-all~e39m1</th>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323318.24-all~e39m1</th>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323362.24-all~e39m1</th>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323441.24-all~e39m1</th>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323466.24-all~e39m1</th>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_-_70323225.22-all~e39m1</th>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_-_70323276.23-all~e39m1</th>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_-_70323332.23-all~e39m1</th>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_-_70323472.23-all~e39m1</th>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>essential_crispri=AARS_+_70323152.24-all~e39m1</th>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>essential_crispri=AARS_+_70323196.24-all~e39m1</th>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>essential_crispri=AARS_+_70323240.24-all~e39m1</th>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>essential_crispri=AARS_+_70323322.23-all~e39m1</th>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>essential_crispri=AARS_+_70323353.24-all~e39m1</th>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>essential_crispri=AARS_+_70323368.23-all~e39m1</th>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>essential_crispri=AARS_-_70323225.23-all~e39m1</th>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>essential_crispri=AARS_-_70323237.23-all~e39m1</th>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS2_+_44280792.21-all~e39m1</th>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS2_+_44281006.23-all~e39m1</th>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    same strand\n",
       "sgId                                                           \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...        False\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...        False\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...        False\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...        False\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...        False\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323466...        False\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_-_70323225...         True\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_-_70323276...         True\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_-_70323332...         True\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_-_70323472...         True\n",
       "essential_crispri=AARS_+_70323152.24-all~e39m1            False\n",
       "essential_crispri=AARS_+_70323196.24-all~e39m1            False\n",
       "essential_crispri=AARS_+_70323240.24-all~e39m1            False\n",
       "essential_crispri=AARS_+_70323322.23-all~e39m1            False\n",
       "essential_crispri=AARS_+_70323353.24-all~e39m1            False\n",
       "essential_crispri=AARS_+_70323368.23-all~e39m1            False\n",
       "essential_crispri=AARS_-_70323225.23-all~e39m1             True\n",
       "essential_crispri=AARS_-_70323237.23-all~e39m1             True\n",
       "Drug_Targets+Kinase_Phosphatase=AARS2_+_4428079...        False\n",
       "Drug_Targets+Kinase_Phosphatase=AARS2_+_4428100...        False"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "strand.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "booleanBaseTable = generateBooleanBaseTable(baseTable)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"10\" halign=\"left\">A</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"10\" halign=\"left\">T</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>-30</th>\n",
       "      <th>-29</th>\n",
       "      <th>-28</th>\n",
       "      <th>-27</th>\n",
       "      <th>-26</th>\n",
       "      <th>-25</th>\n",
       "      <th>-24</th>\n",
       "      <th>-23</th>\n",
       "      <th>-22</th>\n",
       "      <th>-21</th>\n",
       "      <th>...</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sgId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323216.24-all~e39m1</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323296.24-all~e39m1</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323318.24-all~e39m1</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323362.24-all~e39m1</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323441.24-all~e39m1</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 160 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                        A                \\\n",
       "                                                      -30    -29    -28   \n",
       "sgId                                                                      \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...  False  False   True   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...   True  False   True   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...   True  False  False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...  False  False  False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...  False  False  False   \n",
       "\n",
       "                                                                         \\\n",
       "                                                      -27    -26    -25   \n",
       "sgId                                                                      \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...  False  False  False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...  False  False  False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...  False  False  False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...   True   True  False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...  False  False  False   \n",
       "\n",
       "                                                                         \\\n",
       "                                                      -24    -23    -22   \n",
       "sgId                                                                      \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...  False  False  False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...   True  False  False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...  False  False   True   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...   True  False  False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...  False  False  False   \n",
       "\n",
       "                                                           ...      T         \\\n",
       "                                                      -21  ...     0      1    \n",
       "sgId                                                       ...                 \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...  False  ...  False  False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...  False  ...  False   True   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...  False  ...  False  False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...  False  ...  False  False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...  False  ...  False  False   \n",
       "\n",
       "                                                                         \\\n",
       "                                                       2      3      4    \n",
       "sgId                                                                      \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...  False   True  False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...  False  False   True   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...  False  False  False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...  False   True  False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...  False  False  False   \n",
       "\n",
       "                                                                         \\\n",
       "                                                       5      6      7    \n",
       "sgId                                                                      \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...  False  False  False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...  False  False  False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...  False  False   True   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...   True  False  False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...   True   True  False   \n",
       "\n",
       "                                                                  \n",
       "                                                       8      9   \n",
       "sgId                                                              \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...  False  False  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...  False  False  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...  False  False  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...  False  False  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...   True  False  \n",
       "\n",
       "[5 rows x 160 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "booleanBaseTable.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "doubleBaseTable = generateBooleanDoubleBaseTable(baseTable)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MultiIndex(levels=[[(u'A', u'A'), (u'A', u'C'), (u'A', u'G'), (u'A', u'T'), (u'C', u'A'), (u'C', u'C'), (u'C', u'G'), (u'C', u'T'), (u'G', u'A'), (u'G', u'C'), (u'G', u'G'), (u'G', u'T'), (u'T', u'A'), (u'T', u'C'), (u'T', u'G'), (u'T', u'T')], [-30, -29, -28, -27, -26, -25, -24, -23, -22, -21, -20, -19, -18, -17, -16, -15, -14, -13, -12, -11, -10, -9, -8, -7, -6, -5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7]],\n",
       "           codes=[[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15], [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37]])"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "doubleBaseTable.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = [z for z in zip(doubleBaseTable.columns.get_level_values(0),\n",
    "                                 doubleBaseTable.columns.get_level_values(1))]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(('A', 'A'), -30),\n",
       " (('A', 'A'), -29),\n",
       " (('A', 'A'), -28),\n",
       " (('A', 'A'), -27),\n",
       " (('A', 'A'), -26),\n",
       " (('A', 'A'), -25),\n",
       " (('A', 'A'), -24),\n",
       " (('A', 'A'), -23),\n",
       " (('A', 'A'), -22),\n",
       " (('A', 'A'), -21),\n",
       " (('A', 'A'), -20),\n",
       " (('A', 'A'), -19),\n",
       " (('A', 'A'), -18),\n",
       " (('A', 'A'), -17),\n",
       " (('A', 'A'), -16),\n",
       " (('A', 'A'), -15),\n",
       " (('A', 'A'), -14),\n",
       " (('A', 'A'), -13),\n",
       " (('A', 'A'), -12),\n",
       " (('A', 'A'), -11),\n",
       " (('A', 'A'), -10),\n",
       " (('A', 'A'), -9),\n",
       " (('A', 'A'), -8),\n",
       " (('A', 'A'), -7),\n",
       " (('A', 'A'), -6),\n",
       " (('A', 'A'), -5),\n",
       " (('A', 'A'), -4),\n",
       " (('A', 'A'), -3),\n",
       " (('A', 'A'), -2),\n",
       " (('A', 'A'), -1),\n",
       " (('A', 'A'), 0),\n",
       " (('A', 'A'), 1),\n",
       " (('A', 'A'), 2),\n",
       " (('A', 'A'), 3),\n",
       " (('A', 'A'), 4),\n",
       " (('A', 'A'), 5),\n",
       " (('A', 'A'), 6),\n",
       " (('A', 'A'), 7),\n",
       " (('A', 'G'), -30),\n",
       " (('A', 'G'), -29),\n",
       " (('A', 'G'), -28),\n",
       " (('A', 'G'), -27),\n",
       " (('A', 'G'), -26),\n",
       " (('A', 'G'), -25),\n",
       " (('A', 'G'), -24),\n",
       " (('A', 'G'), -23),\n",
       " (('A', 'G'), -22),\n",
       " (('A', 'G'), -21),\n",
       " (('A', 'G'), -20),\n",
       " (('A', 'G'), -19),\n",
       " (('A', 'G'), -18),\n",
       " (('A', 'G'), -17),\n",
       " (('A', 'G'), -16),\n",
       " (('A', 'G'), -15),\n",
       " (('A', 'G'), -14),\n",
       " (('A', 'G'), -13),\n",
       " (('A', 'G'), -12),\n",
       " (('A', 'G'), -11),\n",
       " (('A', 'G'), -10),\n",
       " (('A', 'G'), -9),\n",
       " (('A', 'G'), -8),\n",
       " (('A', 'G'), -7),\n",
       " (('A', 'G'), -6),\n",
       " (('A', 'G'), -5),\n",
       " (('A', 'G'), -4),\n",
       " (('A', 'G'), -3),\n",
       " (('A', 'G'), -2),\n",
       " (('A', 'G'), -1),\n",
       " (('A', 'G'), 0),\n",
       " (('A', 'G'), 1),\n",
       " (('A', 'G'), 2),\n",
       " (('A', 'G'), 3),\n",
       " (('A', 'G'), 4),\n",
       " (('A', 'G'), 5),\n",
       " (('A', 'G'), 6),\n",
       " (('A', 'G'), 7),\n",
       " (('A', 'C'), -30),\n",
       " (('A', 'C'), -29),\n",
       " (('A', 'C'), -28),\n",
       " (('A', 'C'), -27),\n",
       " (('A', 'C'), -26),\n",
       " (('A', 'C'), -25),\n",
       " (('A', 'C'), -24),\n",
       " (('A', 'C'), -23),\n",
       " (('A', 'C'), -22),\n",
       " (('A', 'C'), -21),\n",
       " (('A', 'C'), -20),\n",
       " (('A', 'C'), -19),\n",
       " (('A', 'C'), -18),\n",
       " (('A', 'C'), -17),\n",
       " (('A', 'C'), -16),\n",
       " (('A', 'C'), -15),\n",
       " (('A', 'C'), -14),\n",
       " (('A', 'C'), -13),\n",
       " (('A', 'C'), -12),\n",
       " (('A', 'C'), -11),\n",
       " (('A', 'C'), -10),\n",
       " (('A', 'C'), -9),\n",
       " (('A', 'C'), -8),\n",
       " (('A', 'C'), -7),\n",
       " (('A', 'C'), -6),\n",
       " (('A', 'C'), -5),\n",
       " (('A', 'C'), -4),\n",
       " (('A', 'C'), -3),\n",
       " (('A', 'C'), -2),\n",
       " (('A', 'C'), -1),\n",
       " (('A', 'C'), 0),\n",
       " (('A', 'C'), 1),\n",
       " (('A', 'C'), 2),\n",
       " (('A', 'C'), 3),\n",
       " (('A', 'C'), 4),\n",
       " (('A', 'C'), 5),\n",
       " (('A', 'C'), 6),\n",
       " (('A', 'C'), 7),\n",
       " (('A', 'T'), -30),\n",
       " (('A', 'T'), -29),\n",
       " (('A', 'T'), -28),\n",
       " (('A', 'T'), -27),\n",
       " (('A', 'T'), -26),\n",
       " (('A', 'T'), -25),\n",
       " (('A', 'T'), -24),\n",
       " (('A', 'T'), -23),\n",
       " (('A', 'T'), -22),\n",
       " (('A', 'T'), -21),\n",
       " (('A', 'T'), -20),\n",
       " (('A', 'T'), -19),\n",
       " (('A', 'T'), -18),\n",
       " (('A', 'T'), -17),\n",
       " (('A', 'T'), -16),\n",
       " (('A', 'T'), -15),\n",
       " (('A', 'T'), -14),\n",
       " (('A', 'T'), -13),\n",
       " (('A', 'T'), -12),\n",
       " (('A', 'T'), -11),\n",
       " (('A', 'T'), -10),\n",
       " (('A', 'T'), -9),\n",
       " (('A', 'T'), -8),\n",
       " (('A', 'T'), -7),\n",
       " (('A', 'T'), -6),\n",
       " (('A', 'T'), -5),\n",
       " (('A', 'T'), -4),\n",
       " (('A', 'T'), -3),\n",
       " (('A', 'T'), -2),\n",
       " (('A', 'T'), -1),\n",
       " (('A', 'T'), 0),\n",
       " (('A', 'T'), 1),\n",
       " (('A', 'T'), 2),\n",
       " (('A', 'T'), 3),\n",
       " (('A', 'T'), 4),\n",
       " (('A', 'T'), 5),\n",
       " (('A', 'T'), 6),\n",
       " (('A', 'T'), 7),\n",
       " (('G', 'A'), -30),\n",
       " (('G', 'A'), -29),\n",
       " (('G', 'A'), -28),\n",
       " (('G', 'A'), -27),\n",
       " (('G', 'A'), -26),\n",
       " (('G', 'A'), -25),\n",
       " (('G', 'A'), -24),\n",
       " (('G', 'A'), -23),\n",
       " (('G', 'A'), -22),\n",
       " (('G', 'A'), -21),\n",
       " (('G', 'A'), -20),\n",
       " (('G', 'A'), -19),\n",
       " (('G', 'A'), -18),\n",
       " (('G', 'A'), -17),\n",
       " (('G', 'A'), -16),\n",
       " (('G', 'A'), -15),\n",
       " (('G', 'A'), -14),\n",
       " (('G', 'A'), -13),\n",
       " (('G', 'A'), -12),\n",
       " (('G', 'A'), -11),\n",
       " (('G', 'A'), -10),\n",
       " (('G', 'A'), -9),\n",
       " (('G', 'A'), -8),\n",
       " (('G', 'A'), -7),\n",
       " (('G', 'A'), -6),\n",
       " (('G', 'A'), -5),\n",
       " (('G', 'A'), -4),\n",
       " (('G', 'A'), -3),\n",
       " (('G', 'A'), -2),\n",
       " (('G', 'A'), -1),\n",
       " (('G', 'A'), 0),\n",
       " (('G', 'A'), 1),\n",
       " (('G', 'A'), 2),\n",
       " (('G', 'A'), 3),\n",
       " (('G', 'A'), 4),\n",
       " (('G', 'A'), 5),\n",
       " (('G', 'A'), 6),\n",
       " (('G', 'A'), 7),\n",
       " (('G', 'G'), -30),\n",
       " (('G', 'G'), -29),\n",
       " (('G', 'G'), -28),\n",
       " (('G', 'G'), -27),\n",
       " (('G', 'G'), -26),\n",
       " (('G', 'G'), -25),\n",
       " (('G', 'G'), -24),\n",
       " (('G', 'G'), -23),\n",
       " (('G', 'G'), -22),\n",
       " (('G', 'G'), -21),\n",
       " (('G', 'G'), -20),\n",
       " (('G', 'G'), -19),\n",
       " (('G', 'G'), -18),\n",
       " (('G', 'G'), -17),\n",
       " (('G', 'G'), -16),\n",
       " (('G', 'G'), -15),\n",
       " (('G', 'G'), -14),\n",
       " (('G', 'G'), -13),\n",
       " (('G', 'G'), -12),\n",
       " (('G', 'G'), -11),\n",
       " (('G', 'G'), -10),\n",
       " (('G', 'G'), -9),\n",
       " (('G', 'G'), -8),\n",
       " (('G', 'G'), -7),\n",
       " (('G', 'G'), -6),\n",
       " (('G', 'G'), -5),\n",
       " (('G', 'G'), -4),\n",
       " (('G', 'G'), -3),\n",
       " (('G', 'G'), -2),\n",
       " (('G', 'G'), -1),\n",
       " (('G', 'G'), 0),\n",
       " (('G', 'G'), 1),\n",
       " (('G', 'G'), 2),\n",
       " (('G', 'G'), 3),\n",
       " (('G', 'G'), 4),\n",
       " (('G', 'G'), 5),\n",
       " (('G', 'G'), 6),\n",
       " (('G', 'G'), 7),\n",
       " (('G', 'C'), -30),\n",
       " (('G', 'C'), -29),\n",
       " (('G', 'C'), -28),\n",
       " (('G', 'C'), -27),\n",
       " (('G', 'C'), -26),\n",
       " (('G', 'C'), -25),\n",
       " (('G', 'C'), -24),\n",
       " (('G', 'C'), -23),\n",
       " (('G', 'C'), -22),\n",
       " (('G', 'C'), -21),\n",
       " (('G', 'C'), -20),\n",
       " (('G', 'C'), -19),\n",
       " (('G', 'C'), -18),\n",
       " (('G', 'C'), -17),\n",
       " (('G', 'C'), -16),\n",
       " (('G', 'C'), -15),\n",
       " (('G', 'C'), -14),\n",
       " (('G', 'C'), -13),\n",
       " (('G', 'C'), -12),\n",
       " (('G', 'C'), -11),\n",
       " (('G', 'C'), -10),\n",
       " (('G', 'C'), -9),\n",
       " (('G', 'C'), -8),\n",
       " (('G', 'C'), -7),\n",
       " (('G', 'C'), -6),\n",
       " (('G', 'C'), -5),\n",
       " (('G', 'C'), -4),\n",
       " (('G', 'C'), -3),\n",
       " (('G', 'C'), -2),\n",
       " (('G', 'C'), -1),\n",
       " (('G', 'C'), 0),\n",
       " (('G', 'C'), 1),\n",
       " (('G', 'C'), 2),\n",
       " (('G', 'C'), 3),\n",
       " (('G', 'C'), 4),\n",
       " (('G', 'C'), 5),\n",
       " (('G', 'C'), 6),\n",
       " (('G', 'C'), 7),\n",
       " (('G', 'T'), -30),\n",
       " (('G', 'T'), -29),\n",
       " (('G', 'T'), -28),\n",
       " (('G', 'T'), -27),\n",
       " (('G', 'T'), -26),\n",
       " (('G', 'T'), -25),\n",
       " (('G', 'T'), -24),\n",
       " (('G', 'T'), -23),\n",
       " (('G', 'T'), -22),\n",
       " (('G', 'T'), -21),\n",
       " (('G', 'T'), -20),\n",
       " (('G', 'T'), -19),\n",
       " (('G', 'T'), -18),\n",
       " (('G', 'T'), -17),\n",
       " (('G', 'T'), -16),\n",
       " (('G', 'T'), -15),\n",
       " (('G', 'T'), -14),\n",
       " (('G', 'T'), -13),\n",
       " (('G', 'T'), -12),\n",
       " (('G', 'T'), -11),\n",
       " (('G', 'T'), -10),\n",
       " (('G', 'T'), -9),\n",
       " (('G', 'T'), -8),\n",
       " (('G', 'T'), -7),\n",
       " (('G', 'T'), -6),\n",
       " (('G', 'T'), -5),\n",
       " (('G', 'T'), -4),\n",
       " (('G', 'T'), -3),\n",
       " (('G', 'T'), -2),\n",
       " (('G', 'T'), -1),\n",
       " (('G', 'T'), 0),\n",
       " (('G', 'T'), 1),\n",
       " (('G', 'T'), 2),\n",
       " (('G', 'T'), 3),\n",
       " (('G', 'T'), 4),\n",
       " (('G', 'T'), 5),\n",
       " (('G', 'T'), 6),\n",
       " (('G', 'T'), 7),\n",
       " (('C', 'A'), -30),\n",
       " (('C', 'A'), -29),\n",
       " (('C', 'A'), -28),\n",
       " (('C', 'A'), -27),\n",
       " (('C', 'A'), -26),\n",
       " (('C', 'A'), -25),\n",
       " (('C', 'A'), -24),\n",
       " (('C', 'A'), -23),\n",
       " (('C', 'A'), -22),\n",
       " (('C', 'A'), -21),\n",
       " (('C', 'A'), -20),\n",
       " (('C', 'A'), -19),\n",
       " (('C', 'A'), -18),\n",
       " (('C', 'A'), -17),\n",
       " (('C', 'A'), -16),\n",
       " (('C', 'A'), -15),\n",
       " (('C', 'A'), -14),\n",
       " (('C', 'A'), -13),\n",
       " (('C', 'A'), -12),\n",
       " (('C', 'A'), -11),\n",
       " (('C', 'A'), -10),\n",
       " (('C', 'A'), -9),\n",
       " (('C', 'A'), -8),\n",
       " (('C', 'A'), -7),\n",
       " (('C', 'A'), -6),\n",
       " (('C', 'A'), -5),\n",
       " (('C', 'A'), -4),\n",
       " (('C', 'A'), -3),\n",
       " (('C', 'A'), -2),\n",
       " (('C', 'A'), -1),\n",
       " (('C', 'A'), 0),\n",
       " (('C', 'A'), 1),\n",
       " (('C', 'A'), 2),\n",
       " (('C', 'A'), 3),\n",
       " (('C', 'A'), 4),\n",
       " (('C', 'A'), 5),\n",
       " (('C', 'A'), 6),\n",
       " (('C', 'A'), 7),\n",
       " (('C', 'G'), -30),\n",
       " (('C', 'G'), -29),\n",
       " (('C', 'G'), -28),\n",
       " (('C', 'G'), -27),\n",
       " (('C', 'G'), -26),\n",
       " (('C', 'G'), -25),\n",
       " (('C', 'G'), -24),\n",
       " (('C', 'G'), -23),\n",
       " (('C', 'G'), -22),\n",
       " (('C', 'G'), -21),\n",
       " (('C', 'G'), -20),\n",
       " (('C', 'G'), -19),\n",
       " (('C', 'G'), -18),\n",
       " (('C', 'G'), -17),\n",
       " (('C', 'G'), -16),\n",
       " (('C', 'G'), -15),\n",
       " (('C', 'G'), -14),\n",
       " (('C', 'G'), -13),\n",
       " (('C', 'G'), -12),\n",
       " (('C', 'G'), -11),\n",
       " (('C', 'G'), -10),\n",
       " (('C', 'G'), -9),\n",
       " (('C', 'G'), -8),\n",
       " (('C', 'G'), -7),\n",
       " (('C', 'G'), -6),\n",
       " (('C', 'G'), -5),\n",
       " (('C', 'G'), -4),\n",
       " (('C', 'G'), -3),\n",
       " (('C', 'G'), -2),\n",
       " (('C', 'G'), -1),\n",
       " (('C', 'G'), 0),\n",
       " (('C', 'G'), 1),\n",
       " (('C', 'G'), 2),\n",
       " (('C', 'G'), 3),\n",
       " (('C', 'G'), 4),\n",
       " (('C', 'G'), 5),\n",
       " (('C', 'G'), 6),\n",
       " (('C', 'G'), 7),\n",
       " (('C', 'C'), -30),\n",
       " (('C', 'C'), -29),\n",
       " (('C', 'C'), -28),\n",
       " (('C', 'C'), -27),\n",
       " (('C', 'C'), -26),\n",
       " (('C', 'C'), -25),\n",
       " (('C', 'C'), -24),\n",
       " (('C', 'C'), -23),\n",
       " (('C', 'C'), -22),\n",
       " (('C', 'C'), -21),\n",
       " (('C', 'C'), -20),\n",
       " (('C', 'C'), -19),\n",
       " (('C', 'C'), -18),\n",
       " (('C', 'C'), -17),\n",
       " (('C', 'C'), -16),\n",
       " (('C', 'C'), -15),\n",
       " (('C', 'C'), -14),\n",
       " (('C', 'C'), -13),\n",
       " (('C', 'C'), -12),\n",
       " (('C', 'C'), -11),\n",
       " (('C', 'C'), -10),\n",
       " (('C', 'C'), -9),\n",
       " (('C', 'C'), -8),\n",
       " (('C', 'C'), -7),\n",
       " (('C', 'C'), -6),\n",
       " (('C', 'C'), -5),\n",
       " (('C', 'C'), -4),\n",
       " (('C', 'C'), -3),\n",
       " (('C', 'C'), -2),\n",
       " (('C', 'C'), -1),\n",
       " (('C', 'C'), 0),\n",
       " (('C', 'C'), 1),\n",
       " (('C', 'C'), 2),\n",
       " (('C', 'C'), 3),\n",
       " (('C', 'C'), 4),\n",
       " (('C', 'C'), 5),\n",
       " (('C', 'C'), 6),\n",
       " (('C', 'C'), 7),\n",
       " (('C', 'T'), -30),\n",
       " (('C', 'T'), -29),\n",
       " (('C', 'T'), -28),\n",
       " (('C', 'T'), -27),\n",
       " (('C', 'T'), -26),\n",
       " (('C', 'T'), -25),\n",
       " (('C', 'T'), -24),\n",
       " (('C', 'T'), -23),\n",
       " (('C', 'T'), -22),\n",
       " (('C', 'T'), -21),\n",
       " (('C', 'T'), -20),\n",
       " (('C', 'T'), -19),\n",
       " (('C', 'T'), -18),\n",
       " (('C', 'T'), -17),\n",
       " (('C', 'T'), -16),\n",
       " (('C', 'T'), -15),\n",
       " (('C', 'T'), -14),\n",
       " (('C', 'T'), -13),\n",
       " (('C', 'T'), -12),\n",
       " (('C', 'T'), -11),\n",
       " (('C', 'T'), -10),\n",
       " (('C', 'T'), -9),\n",
       " (('C', 'T'), -8),\n",
       " (('C', 'T'), -7),\n",
       " (('C', 'T'), -6),\n",
       " (('C', 'T'), -5),\n",
       " (('C', 'T'), -4),\n",
       " (('C', 'T'), -3),\n",
       " (('C', 'T'), -2),\n",
       " (('C', 'T'), -1),\n",
       " (('C', 'T'), 0),\n",
       " (('C', 'T'), 1),\n",
       " (('C', 'T'), 2),\n",
       " (('C', 'T'), 3),\n",
       " (('C', 'T'), 4),\n",
       " (('C', 'T'), 5),\n",
       " (('C', 'T'), 6),\n",
       " (('C', 'T'), 7),\n",
       " (('T', 'A'), -30),\n",
       " (('T', 'A'), -29),\n",
       " (('T', 'A'), -28),\n",
       " (('T', 'A'), -27),\n",
       " (('T', 'A'), -26),\n",
       " (('T', 'A'), -25),\n",
       " (('T', 'A'), -24),\n",
       " (('T', 'A'), -23),\n",
       " (('T', 'A'), -22),\n",
       " (('T', 'A'), -21),\n",
       " (('T', 'A'), -20),\n",
       " (('T', 'A'), -19),\n",
       " (('T', 'A'), -18),\n",
       " (('T', 'A'), -17),\n",
       " (('T', 'A'), -16),\n",
       " (('T', 'A'), -15),\n",
       " (('T', 'A'), -14),\n",
       " (('T', 'A'), -13),\n",
       " (('T', 'A'), -12),\n",
       " (('T', 'A'), -11),\n",
       " (('T', 'A'), -10),\n",
       " (('T', 'A'), -9),\n",
       " (('T', 'A'), -8),\n",
       " (('T', 'A'), -7),\n",
       " (('T', 'A'), -6),\n",
       " (('T', 'A'), -5),\n",
       " (('T', 'A'), -4),\n",
       " (('T', 'A'), -3),\n",
       " (('T', 'A'), -2),\n",
       " (('T', 'A'), -1),\n",
       " (('T', 'A'), 0),\n",
       " (('T', 'A'), 1),\n",
       " (('T', 'A'), 2),\n",
       " (('T', 'A'), 3),\n",
       " (('T', 'A'), 4),\n",
       " (('T', 'A'), 5),\n",
       " (('T', 'A'), 6),\n",
       " (('T', 'A'), 7),\n",
       " (('T', 'G'), -30),\n",
       " (('T', 'G'), -29),\n",
       " (('T', 'G'), -28),\n",
       " (('T', 'G'), -27),\n",
       " (('T', 'G'), -26),\n",
       " (('T', 'G'), -25),\n",
       " (('T', 'G'), -24),\n",
       " (('T', 'G'), -23),\n",
       " (('T', 'G'), -22),\n",
       " (('T', 'G'), -21),\n",
       " (('T', 'G'), -20),\n",
       " (('T', 'G'), -19),\n",
       " (('T', 'G'), -18),\n",
       " (('T', 'G'), -17),\n",
       " (('T', 'G'), -16),\n",
       " (('T', 'G'), -15),\n",
       " (('T', 'G'), -14),\n",
       " (('T', 'G'), -13),\n",
       " (('T', 'G'), -12),\n",
       " (('T', 'G'), -11),\n",
       " (('T', 'G'), -10),\n",
       " (('T', 'G'), -9),\n",
       " (('T', 'G'), -8),\n",
       " (('T', 'G'), -7),\n",
       " (('T', 'G'), -6),\n",
       " (('T', 'G'), -5),\n",
       " (('T', 'G'), -4),\n",
       " (('T', 'G'), -3),\n",
       " (('T', 'G'), -2),\n",
       " (('T', 'G'), -1),\n",
       " (('T', 'G'), 0),\n",
       " (('T', 'G'), 1),\n",
       " (('T', 'G'), 2),\n",
       " (('T', 'G'), 3),\n",
       " (('T', 'G'), 4),\n",
       " (('T', 'G'), 5),\n",
       " (('T', 'G'), 6),\n",
       " (('T', 'G'), 7),\n",
       " (('T', 'C'), -30),\n",
       " (('T', 'C'), -29),\n",
       " (('T', 'C'), -28),\n",
       " (('T', 'C'), -27),\n",
       " (('T', 'C'), -26),\n",
       " (('T', 'C'), -25),\n",
       " (('T', 'C'), -24),\n",
       " (('T', 'C'), -23),\n",
       " (('T', 'C'), -22),\n",
       " (('T', 'C'), -21),\n",
       " (('T', 'C'), -20),\n",
       " (('T', 'C'), -19),\n",
       " (('T', 'C'), -18),\n",
       " (('T', 'C'), -17),\n",
       " (('T', 'C'), -16),\n",
       " (('T', 'C'), -15),\n",
       " (('T', 'C'), -14),\n",
       " (('T', 'C'), -13),\n",
       " (('T', 'C'), -12),\n",
       " (('T', 'C'), -11),\n",
       " (('T', 'C'), -10),\n",
       " (('T', 'C'), -9),\n",
       " (('T', 'C'), -8),\n",
       " (('T', 'C'), -7),\n",
       " (('T', 'C'), -6),\n",
       " (('T', 'C'), -5),\n",
       " (('T', 'C'), -4),\n",
       " (('T', 'C'), -3),\n",
       " (('T', 'C'), -2),\n",
       " (('T', 'C'), -1),\n",
       " (('T', 'C'), 0),\n",
       " (('T', 'C'), 1),\n",
       " (('T', 'C'), 2),\n",
       " (('T', 'C'), 3),\n",
       " (('T', 'C'), 4),\n",
       " (('T', 'C'), 5),\n",
       " (('T', 'C'), 6),\n",
       " (('T', 'C'), 7),\n",
       " (('T', 'T'), -30),\n",
       " (('T', 'T'), -29),\n",
       " (('T', 'T'), -28),\n",
       " (('T', 'T'), -27),\n",
       " (('T', 'T'), -26),\n",
       " (('T', 'T'), -25),\n",
       " (('T', 'T'), -24),\n",
       " (('T', 'T'), -23),\n",
       " (('T', 'T'), -22),\n",
       " (('T', 'T'), -21),\n",
       " (('T', 'T'), -20),\n",
       " (('T', 'T'), -19),\n",
       " (('T', 'T'), -18),\n",
       " (('T', 'T'), -17),\n",
       " (('T', 'T'), -16),\n",
       " (('T', 'T'), -15),\n",
       " (('T', 'T'), -14),\n",
       " (('T', 'T'), -13),\n",
       " (('T', 'T'), -12),\n",
       " (('T', 'T'), -11),\n",
       " (('T', 'T'), -10),\n",
       " (('T', 'T'), -9),\n",
       " (('T', 'T'), -8),\n",
       " (('T', 'T'), -7),\n",
       " (('T', 'T'), -6),\n",
       " (('T', 'T'), -5),\n",
       " (('T', 'T'), -4),\n",
       " (('T', 'T'), -3),\n",
       " (('T', 'T'), -2),\n",
       " (('T', 'T'), -1),\n",
       " (('T', 'T'), 0),\n",
       " (('T', 'T'), 1),\n",
       " (('T', 'T'), 2),\n",
       " (('T', 'T'), 3),\n",
       " (('T', 'T'), 4),\n",
       " (('T', 'T'), 5),\n",
       " (('T', 'T'), 6),\n",
       " (('T', 'T'), 7)]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cols = [' '.join(z) for z in zip([\"\".join(x) for x in doubleBaseTable.columns.get_level_values(0)],\n",
    "#                                                     [str(y) for y in doubleBaseTable.columns.get_level_values(1)])]\n",
    "doubleBaseTable.columns = doubleBaseTable.columns.droplevel(level=1)\n",
    "doubleBaseTable.columns = cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>((A, A), -30)</th>\n",
       "      <th>((A, A), -29)</th>\n",
       "      <th>((A, A), -28)</th>\n",
       "      <th>((A, A), -27)</th>\n",
       "      <th>((A, A), -26)</th>\n",
       "      <th>((A, A), -25)</th>\n",
       "      <th>((A, A), -24)</th>\n",
       "      <th>((A, A), -23)</th>\n",
       "      <th>((A, A), -22)</th>\n",
       "      <th>((A, A), -21)</th>\n",
       "      <th>...</th>\n",
       "      <th>((T, T), -2)</th>\n",
       "      <th>((T, T), -1)</th>\n",
       "      <th>((T, T), 0)</th>\n",
       "      <th>((T, T), 1)</th>\n",
       "      <th>((T, T), 2)</th>\n",
       "      <th>((T, T), 3)</th>\n",
       "      <th>((T, T), 4)</th>\n",
       "      <th>((T, T), 5)</th>\n",
       "      <th>((T, T), 6)</th>\n",
       "      <th>((T, T), 7)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sgId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323216.24-all~e39m1</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323296.24-all~e39m1</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323318.24-all~e39m1</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323362.24-all~e39m1</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323441.24-all~e39m1</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 608 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    ((A, A), -30)  \\\n",
       "sgId                                                                \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...          False   \n",
       "\n",
       "                                                    ((A, A), -29)  \\\n",
       "sgId                                                                \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...          False   \n",
       "\n",
       "                                                    ((A, A), -28)  \\\n",
       "sgId                                                                \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...          False   \n",
       "\n",
       "                                                    ((A, A), -27)  \\\n",
       "sgId                                                                \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...           True   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...          False   \n",
       "\n",
       "                                                    ((A, A), -26)  \\\n",
       "sgId                                                                \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...          False   \n",
       "\n",
       "                                                    ((A, A), -25)  \\\n",
       "sgId                                                                \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...          False   \n",
       "\n",
       "                                                    ((A, A), -24)  \\\n",
       "sgId                                                                \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...          False   \n",
       "\n",
       "                                                    ((A, A), -23)  \\\n",
       "sgId                                                                \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...          False   \n",
       "\n",
       "                                                    ((A, A), -22)  \\\n",
       "sgId                                                                \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...          False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...          False   \n",
       "\n",
       "                                                    ((A, A), -21)  ...  \\\n",
       "sgId                                                               ...   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...          False  ...   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...          False  ...   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...          False  ...   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...          False  ...   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...          False  ...   \n",
       "\n",
       "                                                    ((T, T), -2)  \\\n",
       "sgId                                                               \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...         False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...         False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...         False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...         False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...         False   \n",
       "\n",
       "                                                    ((T, T), -1)  ((T, T), 0)  \\\n",
       "sgId                                                                            \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...         False        False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...         False        False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...         False        False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...         False        False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...         False        False   \n",
       "\n",
       "                                                    ((T, T), 1)  ((T, T), 2)  \\\n",
       "sgId                                                                           \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...        False        False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...        False        False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...        False        False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...        False        False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...        False        False   \n",
       "\n",
       "                                                    ((T, T), 3)  ((T, T), 4)  \\\n",
       "sgId                                                                           \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...        False        False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...        False        False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...        False        False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...        False        False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...        False        False   \n",
       "\n",
       "                                                    ((T, T), 5)  ((T, T), 6)  \\\n",
       "sgId                                                                           \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...        False        False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...        False        False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...        False        False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...        False        False   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...         True        False   \n",
       "\n",
       "                                                    ((T, T), 7)  \n",
       "sgId                                                             \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...        False  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...        False  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...        False  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...        False  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...        False  \n",
       "\n",
       "[5 rows x 608 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "doubleBaseTable.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "baseList = ['A','G','C','T']\n",
    "homopolymerTable = pd.concat([libraryTable.apply(lambda row: np.floor(getMaxLengthHomopolymer(row['sequence'], base)), axis=1) for base in baseList],keys=baseList,axis=1)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>G</th>\n",
       "      <th>C</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sgId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323216.24-all~e39m1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323296.24-all~e39m1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323318.24-all~e39m1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323362.24-all~e39m1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323441.24-all~e39m1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      A    G    C    T\n",
       "sgId                                                                  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...  1.0  2.0  4.0  1.0\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...  1.0  2.0  3.0  1.0\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...  1.0  2.0  3.0  1.0\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...  1.0  3.0  3.0  2.0\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...  1.0  2.0  1.0  2.0"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "homopolymerTable.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "baseFractions = pd.concat([libraryTable.apply(lambda row: getFractionBaseList(row['sequence'], ['A']),axis=1),\n",
    "                        libraryTable.apply(lambda row: getFractionBaseList(row['sequence'], ['G']),axis=1),\n",
    "                        libraryTable.apply(lambda row: getFractionBaseList(row['sequence'], ['C']),axis=1),\n",
    "                        libraryTable.apply(lambda row: getFractionBaseList(row['sequence'], ['T']),axis=1),\n",
    "                        libraryTable.apply(lambda row: getFractionBaseList(row['sequence'], ['G','C']),axis=1),\n",
    "                        libraryTable.apply(lambda row: getFractionBaseList(row['sequence'], ['G','A']),axis=1),\n",
    "                        libraryTable.apply(lambda row: getFractionBaseList(row['sequence'], ['C','A']),axis=1)],keys=['A','G','C','T','GC','purine','CA'],axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>G</th>\n",
       "      <th>C</th>\n",
       "      <th>T</th>\n",
       "      <th>GC</th>\n",
       "      <th>purine</th>\n",
       "      <th>CA</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sgId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323216.24-all~e39m1</th>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.809524</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.619048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323296.24-all~e39m1</th>\n",
       "      <td>0.095238</td>\n",
       "      <td>0.380952</td>\n",
       "      <td>0.380952</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.761905</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.476190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323318.24-all~e39m1</th>\n",
       "      <td>0.238095</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.238095</td>\n",
       "      <td>0.095238</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.476190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323362.24-all~e39m1</th>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.523810</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.809524</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug_Targets+Kinase_Phosphatase=AARS_+_70323441.24-all~e39m1</th>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.238095</td>\n",
       "      <td>0.190476</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>0.380952</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           A         G  \\\n",
       "sgId                                                                     \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...  0.142857  0.333333   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...  0.095238  0.380952   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...  0.238095  0.428571   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...  0.047619  0.523810   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...  0.142857  0.428571   \n",
       "\n",
       "                                                           C         T  \\\n",
       "sgId                                                                     \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...  0.476190  0.047619   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...  0.380952  0.142857   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...  0.238095  0.095238   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...  0.285714  0.142857   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...  0.238095  0.190476   \n",
       "\n",
       "                                                          GC    purine  \\\n",
       "sgId                                                                     \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...  0.809524  0.476190   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...  0.761905  0.476190   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...  0.666667  0.666667   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...  0.809524  0.571429   \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...  0.666667  0.571429   \n",
       "\n",
       "                                                          CA  \n",
       "sgId                                                          \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216...  0.619048  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296...  0.476190  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318...  0.476190  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362...  0.333333  \n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441...  0.380952  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "baseFractions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "dnaseSeries = getChromatinDataSeriesByGene(bwFileHandleDict['dnase'], libraryTable, sgInfoTable, p1p2Table, sgrnaPositionTable_p1p2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sgId\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216.24-all~e39m1    0.047101\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296.24-all~e39m1    0.038820\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318.24-all~e39m1    0.131211\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362.24-all~e39m1    0.239001\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441.24-all~e39m1    0.190864\n",
       "dtype: float64"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dnaseSeries.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "faireSeries = getChromatinDataSeriesByGene(bwFileHandleDict['faire'], libraryTable, sgInfoTable, p1p2Table, sgrnaPositionTable_p1p2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sgId\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216.24-all~e39m1    0.297187\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296.24-all~e39m1    0.350605\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318.24-all~e39m1    0.391916\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362.24-all~e39m1    0.491987\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441.24-all~e39m1    0.812678\n",
       "dtype: float64"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "faireSeries.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<string>:619: RuntimeWarning: Mean of empty slice\n",
      "<string>:595: RuntimeWarning: All-NaN slice encountered\n",
      "<string>:619: RuntimeWarning: invalid value encountered in float_scalars\n"
     ]
    }
   ],
   "source": [
    "mnaseSeries = getChromatinDataSeriesByGene(bwFileHandleDict['mnase'], libraryTable, sgInfoTable, p1p2Table, sgrnaPositionTable_p1p2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sgId\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323216.24-all~e39m1    0.059896\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323296.24-all~e39m1    0.101563\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323318.24-all~e39m1    0.023438\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323362.24-all~e39m1    0.033854\n",
       "Drug_Targets+Kinase_Phosphatase=AARS_+_70323441.24-all~e39m1    0.079427\n",
       "dtype: float64"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mnaseSeries.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Length of passed values is 0, index implies 18380",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-36-788377e0e126>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mrnafolding\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetRNAfoldingTable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlibraryTable\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/Users/ucbbiosciences/analyses/CRISPRiaDesign/sgRNA_learning.py\u001b[0m in \u001b[0;36mgetRNAfoldingTable\u001b[0;34m(libraryTable)\u001b[0m\n\u001b[1;32m    514\u001b[0m         \u001b[0msubprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'RNAfold --noPS < %s > %s'\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mtempfile_fa\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtempfile_rnafold\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mshell\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    515\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 516\u001b[0;31m         \u001b[0mmfeSeries_noScaffold\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparseViennaMFE\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtempfile_rnafold\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlibraryTable\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    517\u001b[0m         \u001b[0misPaired\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparseViennaPairing\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtempfile_rnafold\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlibraryTable\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    518\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ucbbiosciences/analyses/CRISPRiaDesign/sgRNA_learning.py\u001b[0m in \u001b[0;36mparseViennaMFE\u001b[0;34m(viennaOutputFile, libraryTable)\u001b[0m\n\u001b[1;32m    540\u001b[0m                         \u001b[0;32mif\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m%\u001b[0m\u001b[0;36m3\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    541\u001b[0m                                 \u001b[0mmfeList\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfloat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstrip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstrip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'.() '\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 542\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSeries\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmfeList\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlibraryTable\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'RNA minimum free energy'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    543\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    544\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mparseViennaPairing\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mviennaOutputFile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlibraryTable\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ucbbiosciences/software/anaconda2/envs/CRISPRiaDesign/lib/python2.7/site-packages/pandas/core/series.pyc\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, data, index, dtype, name, copy, fastpath)\u001b[0m\n\u001b[1;32m    247\u001b[0m                             \u001b[0;34m'Length of passed values is {val}, '\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    248\u001b[0m                             \u001b[0;34m'index implies {ind}'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 249\u001b[0;31m                             .format(val=len(data), ind=len(index)))\n\u001b[0m\u001b[1;32m    250\u001b[0m                 \u001b[0;32mexcept\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    251\u001b[0m                     \u001b[0;32mpass\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Length of passed values is 0, index implies 18380"
     ]
    }
   ],
   "source": [
    "rnafolding = getRNAfoldingTable(libraryTable)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rnafolding.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sgrnaPositionTable_p1p2.iloc[:,2:].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rnafolding.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df1 = pd.concat([dnaseSeries,faireSeries,mnaseSeries],keys=['DNase','FAIRE','MNase'], axis=1)\n",
    "df2 = pd.concat([dnaseSeries,faireSeries,mnaseSeries],keys=['DNase','FAIRE','MNase'], axis=1, sort=False)\n",
    "\n",
    "all(df1.index == df2.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "epigenomic = pd.concat([dnaseSeries,faireSeries,mnaseSeries],keys=['DNase','FAIRE','MNase'], axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "df_dict = {\n",
    "    'length': lengthSeries,\n",
    "    'distance': sgrnaPositionTable_p1p2.iloc[:,2:],\n",
    "    'homopolymers': homopolymerTable,\n",
    "    'base fractions': baseFractions,\n",
    "    'strand': strand,\n",
    "    'base table-A': booleanBaseTable['A'],\n",
    "    'base table-T': booleanBaseTable['T'],\n",
    "    'base table-G': booleanBaseTable['G'],\n",
    "    'base table-C': booleanBaseTable['C'],\n",
    "    'base dimers': doubleBaseTable,\n",
    "    'accessibility': epigenomic,\n",
    "    'RNA folding-no scaffold': rnafolding['no scaffold'],\n",
    "    'RNA folding-with scaffold': rnafolding['with scaffold'],\n",
    "    'RNA folding-pairing no scaffold': rnafolding['is Paired']\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "_keys, _values = zip(*df_dict.items())\n",
    "paramTable_trainingGuides = pd.concat(_values, axis=1, keys=_keys)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "paramTable_trainingGuides.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Fails:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "paramTable_trainingGuides = generateTypicalParamTable(libraryTable_training,sgInfoTable_training, tssTable, p1p2Table, genomeDict, bwhandleDict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fit parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "u'0.24.2'"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load in the 5-fold cross-validation splits used to generate the model\n",
    "import cPickle\n",
    "with open('data_files/CRISPRi_trainingdata_traintestsets.txt') as infile:\n",
    "    geneFold_train, geneFold_test, fitTable = pd.read_pickle(infile, compression=None)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'bin function': <function numpy.lib.function_base.median>,\n",
       " 'bin width': 0.1,\n",
       " 'min edge data': 50}"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fitTable.loc[('base fractions', 'A'),'params']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>params</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>length</th>\n",
       "      <th>length</th>\n",
       "      <td>binnable_onehot</td>\n",
       "      <td>{u'min edge data': 50, u'bin function': &lt;funct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">distance</th>\n",
       "      <th>primary TSS-Up</th>\n",
       "      <td>continuous</td>\n",
       "      <td>{u'C': [0.01, 0.05, 0.1, 0.5], u'gamma': [1e-0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>primary TSS-Down</th>\n",
       "      <td>continuous</td>\n",
       "      <td>{u'C': [0.01, 0.05, 0.1, 0.5], u'gamma': [1e-0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>secondary TSS-Up</th>\n",
       "      <td>continuous</td>\n",
       "      <td>{u'C': [0.01, 0.05, 0.1, 0.5], u'gamma': [1e-0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>secondary TSS-Down</th>\n",
       "      <td>continuous</td>\n",
       "      <td>{u'C': [0.01, 0.05, 0.1, 0.5], u'gamma': [1e-0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">homopolymers</th>\n",
       "      <th>A</th>\n",
       "      <td>binnable_onehot</td>\n",
       "      <td>{u'min edge data': 50, u'bin function': &lt;funct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>binnable_onehot</td>\n",
       "      <td>{u'min edge data': 50, u'bin function': &lt;funct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>binnable_onehot</td>\n",
       "      <td>{u'min edge data': 50, u'bin function': &lt;funct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>binnable_onehot</td>\n",
       "      <td>{u'min edge data': 50, u'bin function': &lt;funct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"7\" valign=\"top\">base fractions</th>\n",
       "      <th>A</th>\n",
       "      <td>binnable_onehot</td>\n",
       "      <td>{u'min edge data': 50, u'bin function': &lt;funct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>binnable_onehot</td>\n",
       "      <td>{u'min edge data': 50, u'bin function': &lt;funct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>binnable_onehot</td>\n",
       "      <td>{u'min edge data': 50, u'bin function': &lt;funct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>binnable_onehot</td>\n",
       "      <td>{u'min edge data': 50, u'bin function': &lt;funct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GC</th>\n",
       "      <td>binnable_onehot</td>\n",
       "      <td>{u'min edge data': 50, u'bin function': &lt;funct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>purine</th>\n",
       "      <td>binnable_onehot</td>\n",
       "      <td>{u'min edge data': 50, u'bin function': &lt;funct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CA</th>\n",
       "      <td>binnable_onehot</td>\n",
       "      <td>{u'min edge data': 50, u'bin function': &lt;funct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>strand</th>\n",
       "      <th>same strand</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"13\" valign=\"top\">base table-A</th>\n",
       "      <th>-30</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-29</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-28</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-27</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-26</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-25</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-24</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-23</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-22</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-21</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-20</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-19</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-18</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"7\" valign=\"top\">base dimers</th>\n",
       "      <th>(('T', 'T'), 1)</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(('T', 'T'), 2)</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(('T', 'T'), 3)</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(('T', 'T'), 4)</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(('T', 'T'), 5)</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(('T', 'T'), 6)</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(('T', 'T'), 7)</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">accessibility</th>\n",
       "      <th>DNase</th>\n",
       "      <td>binnable_onehot</td>\n",
       "      <td>{u'min edge data': 50, u'bin function': &lt;funct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAIRE</th>\n",
       "      <td>binnable_onehot</td>\n",
       "      <td>{u'min edge data': 50, u'bin function': &lt;funct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MNase</th>\n",
       "      <td>binnable_onehot</td>\n",
       "      <td>{u'min edge data': 50, u'bin function': &lt;funct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA folding-no scaffold</th>\n",
       "      <th>RNA minimum free energy</th>\n",
       "      <td>binnable_onehot</td>\n",
       "      <td>{u'min edge data': 50, u'bin function': &lt;funct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA folding-with scaffold</th>\n",
       "      <th>RNA minimum free energy</th>\n",
       "      <td>binnable_onehot</td>\n",
       "      <td>{u'min edge data': 50, u'bin function': &lt;funct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"18\" valign=\"top\">RNA folding-pairing, no scaffold</th>\n",
       "      <th>-20</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-19</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-18</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-17</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-16</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-15</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-14</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-13</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-12</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-11</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-10</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-9</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-8</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-7</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-6</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-5</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-4</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-3</th>\n",
       "      <td>binary</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>808 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                     type  \\\n",
       "length                           length                   binnable_onehot   \n",
       "distance                         primary TSS-Up                continuous   \n",
       "                                 primary TSS-Down              continuous   \n",
       "                                 secondary TSS-Up              continuous   \n",
       "                                 secondary TSS-Down            continuous   \n",
       "homopolymers                     A                        binnable_onehot   \n",
       "                                 G                        binnable_onehot   \n",
       "                                 C                        binnable_onehot   \n",
       "                                 T                        binnable_onehot   \n",
       "base fractions                   A                        binnable_onehot   \n",
       "                                 G                        binnable_onehot   \n",
       "                                 C                        binnable_onehot   \n",
       "                                 T                        binnable_onehot   \n",
       "                                 GC                       binnable_onehot   \n",
       "                                 purine                   binnable_onehot   \n",
       "                                 CA                       binnable_onehot   \n",
       "strand                           same strand                       binary   \n",
       "base table-A                     -30                               binary   \n",
       "                                 -29                               binary   \n",
       "                                 -28                               binary   \n",
       "                                 -27                               binary   \n",
       "                                 -26                               binary   \n",
       "                                 -25                               binary   \n",
       "                                 -24                               binary   \n",
       "                                 -23                               binary   \n",
       "                                 -22                               binary   \n",
       "                                 -21                               binary   \n",
       "                                 -20                               binary   \n",
       "                                 -19                               binary   \n",
       "                                 -18                               binary   \n",
       "...                                                                   ...   \n",
       "base dimers                      (('T', 'T'), 1)                   binary   \n",
       "                                 (('T', 'T'), 2)                   binary   \n",
       "                                 (('T', 'T'), 3)                   binary   \n",
       "                                 (('T', 'T'), 4)                   binary   \n",
       "                                 (('T', 'T'), 5)                   binary   \n",
       "                                 (('T', 'T'), 6)                   binary   \n",
       "                                 (('T', 'T'), 7)                   binary   \n",
       "accessibility                    DNase                    binnable_onehot   \n",
       "                                 FAIRE                    binnable_onehot   \n",
       "                                 MNase                    binnable_onehot   \n",
       "RNA folding-no scaffold          RNA minimum free energy  binnable_onehot   \n",
       "RNA folding-with scaffold        RNA minimum free energy  binnable_onehot   \n",
       "RNA folding-pairing, no scaffold -20                               binary   \n",
       "                                 -19                               binary   \n",
       "                                 -18                               binary   \n",
       "                                 -17                               binary   \n",
       "                                 -16                               binary   \n",
       "                                 -15                               binary   \n",
       "                                 -14                               binary   \n",
       "                                 -13                               binary   \n",
       "                                 -12                               binary   \n",
       "                                 -11                               binary   \n",
       "                                 -10                               binary   \n",
       "                                 -9                                binary   \n",
       "                                 -8                                binary   \n",
       "                                 -7                                binary   \n",
       "                                 -6                                binary   \n",
       "                                 -5                                binary   \n",
       "                                 -4                                binary   \n",
       "                                 -3                                binary   \n",
       "\n",
       "                                                                                                     params  \n",
       "length                           length                   {u'min edge data': 50, u'bin function': <funct...  \n",
       "distance                         primary TSS-Up           {u'C': [0.01, 0.05, 0.1, 0.5], u'gamma': [1e-0...  \n",
       "                                 primary TSS-Down         {u'C': [0.01, 0.05, 0.1, 0.5], u'gamma': [1e-0...  \n",
       "                                 secondary TSS-Up         {u'C': [0.01, 0.05, 0.1, 0.5], u'gamma': [1e-0...  \n",
       "                                 secondary TSS-Down       {u'C': [0.01, 0.05, 0.1, 0.5], u'gamma': [1e-0...  \n",
       "homopolymers                     A                        {u'min edge data': 50, u'bin function': <funct...  \n",
       "                                 G                        {u'min edge data': 50, u'bin function': <funct...  \n",
       "                                 C                        {u'min edge data': 50, u'bin function': <funct...  \n",
       "                                 T                        {u'min edge data': 50, u'bin function': <funct...  \n",
       "base fractions                   A                        {u'min edge data': 50, u'bin function': <funct...  \n",
       "                                 G                        {u'min edge data': 50, u'bin function': <funct...  \n",
       "                                 C                        {u'min edge data': 50, u'bin function': <funct...  \n",
       "                                 T                        {u'min edge data': 50, u'bin function': <funct...  \n",
       "                                 GC                       {u'min edge data': 50, u'bin function': <funct...  \n",
       "                                 purine                   {u'min edge data': 50, u'bin function': <funct...  \n",
       "                                 CA                       {u'min edge data': 50, u'bin function': <funct...  \n",
       "strand                           same strand                                                             {}  \n",
       "base table-A                     -30                                                                     {}  \n",
       "                                 -29                                                                     {}  \n",
       "                                 -28                                                                     {}  \n",
       "                                 -27                                                                     {}  \n",
       "                                 -26                                                                     {}  \n",
       "                                 -25                                                                     {}  \n",
       "                                 -24                                                                     {}  \n",
       "                                 -23                                                                     {}  \n",
       "                                 -22                                                                     {}  \n",
       "                                 -21                                                                     {}  \n",
       "                                 -20                                                                     {}  \n",
       "                                 -19                                                                     {}  \n",
       "                                 -18                                                                     {}  \n",
       "...                                                                                                     ...  \n",
       "base dimers                      (('T', 'T'), 1)                                                         {}  \n",
       "                                 (('T', 'T'), 2)                                                         {}  \n",
       "                                 (('T', 'T'), 3)                                                         {}  \n",
       "                                 (('T', 'T'), 4)                                                         {}  \n",
       "                                 (('T', 'T'), 5)                                                         {}  \n",
       "                                 (('T', 'T'), 6)                                                         {}  \n",
       "                                 (('T', 'T'), 7)                                                         {}  \n",
       "accessibility                    DNase                    {u'min edge data': 50, u'bin function': <funct...  \n",
       "                                 FAIRE                    {u'min edge data': 50, u'bin function': <funct...  \n",
       "                                 MNase                    {u'min edge data': 50, u'bin function': <funct...  \n",
       "RNA folding-no scaffold          RNA minimum free energy  {u'min edge data': 50, u'bin function': <funct...  \n",
       "RNA folding-with scaffold        RNA minimum free energy  {u'min edge data': 50, u'bin function': <funct...  \n",
       "RNA folding-pairing, no scaffold -20                                                                     {}  \n",
       "                                 -19                                                                     {}  \n",
       "                                 -18                                                                     {}  \n",
       "                                 -17                                                                     {}  \n",
       "                                 -16                                                                     {}  \n",
       "                                 -15                                                                     {}  \n",
       "                                 -14                                                                     {}  \n",
       "                                 -13                                                                     {}  \n",
       "                                 -12                                                                     {}  \n",
       "                                 -11                                                                     {}  \n",
       "                                 -10                                                                     {}  \n",
       "                                 -9                                                                      {}  \n",
       "                                 -8                                                                      {}  \n",
       "                                 -7                                                                      {}  \n",
       "                                 -6                                                                      {}  \n",
       "                                 -5                                                                      {}  \n",
       "                                 -4                                                                      {}  \n",
       "                                 -3                                                                      {}  \n",
       "\n",
       "[808 rows x 2 columns]"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fitTable"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### XXX WORKS UNTIL HERE XXX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'C': 0.5, 'gamma': 0.0005}\n",
      "{'C': 0.5, 'gamma': 0.0005}\n",
      "{'C': 0.5, 'gamma': 0.0001}\n",
      "{'C': 0.5, 'gamma': 0.0005}\n",
      "((accessibility, DNase)\n",
      "0.0     14761.0\n",
      "1.0         4.0\n",
      "2.0         1.0\n",
      "3.0         2.0\n",
      "4.0         1.0\n",
      "6.0         1.0\n",
      "7.0         1.0\n",
      "10.0        1.0\n",
      "11.0        1.0\n",
      "Name: (accessibility, DNase), dtype: float64, 0, -10, 50, 1)\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "min edge requirements cannot be met",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-189-5d493b0b93f9>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m transformedParams_train, estimators = fitParams(paramTable_trainingGuides.loc[activityScores.dropna().index].iloc[geneFold_train], \n\u001b[1;32m      2\u001b[0m                                                 \u001b[0mactivityScores\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mactivityScores\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdropna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mgeneFold_train\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m                                                 fitTable)\n\u001b[0m",
      "\u001b[0;32m/Users/ucbbiosciences/analyses/CRISPRiaDesign/sgRNA_learning.py\u001b[0m in \u001b[0;36mfitParams\u001b[0;34m(paramTable, scoreTable, fitTable)\u001b[0m\n\u001b[1;32m    741\u001b[0m                         \u001b[0mparameters\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfitRow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'params'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    742\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 743\u001b[0;31m                         \u001b[0massignedBins\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbinValues\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparameters\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'bin width'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparameters\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'min edge data'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    744\u001b[0m                         \u001b[0mbinGroups\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mscoreTable\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroupby\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0massignedBins\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    745\u001b[0m                         \u001b[0mgroupStats\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbinGroups\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0magg\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparameters\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'bin function'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ucbbiosciences/analyses/CRISPRiaDesign/sgRNA_learning.py\u001b[0m in \u001b[0;36mbinValues\u001b[0;34m(col, binsize, minEdgePoints, edgeOffset)\u001b[0m\n\u001b[1;32m    800\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mi\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbinCounts\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mj\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    801\u001b[0m                         \u001b[0;32mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbinCounts\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mminEdgePoints\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbinsize\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 802\u001b[0;31m                         \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'min edge requirements cannot be met'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    803\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    804\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0medgeOffset\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mNone\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;31m#return strings for bins, fine for grouping, problems for plotting\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: min edge requirements cannot be met"
     ]
    }
   ],
   "source": [
    "transformedParams_train, estimators = fitParams(paramTable_trainingGuides.loc[activityScores.dropna().index].iloc[geneFold_train], \n",
    "                                                activityScores.loc[activityScores.dropna().index].iloc[geneFold_train], \n",
    "                                                fitTable)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "transformedParams_train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('base fractions', 'A') {'C': 0.5, 'gamma': 0.0005}\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-161-e8fb0c732809>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mtransformedParams_train\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mestimators\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfitParams\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparamTable_trainingGuides\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mactivityScores\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdropna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mgeneFold_train\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mactivityScores\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mactivityScores\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdropna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mgeneFold_train\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfitTable\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mtransformedParams_test\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtransformParams\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparamTable_trainingGuides\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mactivityScores\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdropna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mgeneFold_test\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfitTable\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mestimators\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mreg\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlinear_model\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mElasticNetCV\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ml1_ratio\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m.5\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m.75\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m.9\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m.99\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn_jobs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m16\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmax_iter\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m2000\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ucbbiosciences/analyses/CRISPRiaDesign/sgRNA_learning.py\u001b[0m in \u001b[0;36mfitParams\u001b[0;34m(paramTable, scoreTable, fitTable)\u001b[0m\n\u001b[1;32m    718\u001b[0m                         \u001b[0msvr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msvm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSVR\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcache_size\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m500\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    719\u001b[0m                         \u001b[0mclf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgrid_search\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mGridSearchCV\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msvr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparameters\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn_jobs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m16\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverbose\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 720\u001b[0;31m                         \u001b[0mclf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcol_reshape\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mscoreTable\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    721\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    722\u001b[0m                         \u001b[0;32mprint\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mclf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbest_params_\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ucbbiosciences/software/anaconda2/envs/CRISPRiaDesign/lib/python2.7/site-packages/sklearn/grid_search.pyc\u001b[0m in \u001b[0;36mfit\u001b[0;34m(self, X, y)\u001b[0m\n\u001b[1;32m    836\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    837\u001b[0m         \"\"\"\n\u001b[0;32m--> 838\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_fit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mParameterGrid\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparam_grid\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    839\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    840\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ucbbiosciences/software/anaconda2/envs/CRISPRiaDesign/lib/python2.7/site-packages/sklearn/grid_search.pyc\u001b[0m in \u001b[0;36m_fit\u001b[0;34m(self, X, y, parameter_iterable)\u001b[0m\n\u001b[1;32m    572\u001b[0m                                     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit_params\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreturn_parameters\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mTrue\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    573\u001b[0m                                     error_score=self.error_score)\n\u001b[0;32m--> 574\u001b[0;31m                 \u001b[0;32mfor\u001b[0m \u001b[0mparameters\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mparameter_iterable\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    575\u001b[0m                 for train, test in cv)\n\u001b[1;32m    576\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ucbbiosciences/software/anaconda2/envs/CRISPRiaDesign/lib/python2.7/site-packages/sklearn/externals/joblib/parallel.pyc\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, iterable)\u001b[0m\n\u001b[1;32m    787\u001b[0m                 \u001b[0;31m# consumption.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    788\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_iterating\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 789\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mretrieve\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    790\u001b[0m             \u001b[0;31m# Make sure that we get a last message telling us we are done\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    791\u001b[0m             \u001b[0melapsed_time\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_start_time\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ucbbiosciences/software/anaconda2/envs/CRISPRiaDesign/lib/python2.7/site-packages/sklearn/externals/joblib/parallel.pyc\u001b[0m in \u001b[0;36mretrieve\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    697\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    698\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_backend\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'supports_timeout'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 699\u001b[0;31m                     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_output\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mjob\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    700\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    701\u001b[0m                     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_output\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mjob\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ucbbiosciences/software/anaconda2/envs/CRISPRiaDesign/lib/python2.7/multiprocessing/pool.pyc\u001b[0m in \u001b[0;36mget\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    564\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    565\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 566\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    567\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_ready\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    568\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mTimeoutError\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ucbbiosciences/software/anaconda2/envs/CRISPRiaDesign/lib/python2.7/multiprocessing/pool.pyc\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    559\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    560\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_ready\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 561\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_cond\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    562\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    563\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_cond\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrelease\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ucbbiosciences/software/anaconda2/envs/CRISPRiaDesign/lib/python2.7/threading.pyc\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    338\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m    \u001b[0;31m# restore state no matter what (e.g., KeyboardInterrupt)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    339\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mtimeout\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 340\u001b[0;31m                 \u001b[0mwaiter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0macquire\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    341\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0m__debug__\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    342\u001b[0m                     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_note\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"%s.wait(): got it\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "\n",
    "transformedParams_test = transformParams(paramTable_trainingGuides.loc[activityScores.dropna().index].iloc[geneFold_test], fitTable, estimators)\n",
    "\n",
    "reg = linear_model.ElasticNetCV(l1_ratio=[.5, .75, .9, .99, 1], n_jobs=16, max_iter=2000)\n",
    "\n",
    "scaler = preprocessing.StandardScaler()\n",
    "reg.fit(scaler.fit_transform(transformedParams_train), activityScores.loc[activityScores.dropna().index].iloc[geneFold_train])\n",
    "predictedScores = pd.Series(reg.predict(scaler.transform(transformedParams_test)), index=transformedParams_test.index)\n",
    "testScores = activityScores.loc[activityScores.dropna().index].iloc[geneFold_test]\n",
    "\n",
    "print 'Prediction AUC-ROC:', metrics.roc_auc_score((testScores >= .75).values, np.array(predictedScores.values,dtype='float64'))\n",
    "print 'Prediction R^2:', reg.score(scaler.transform(transformedParams_test), testScores)\n",
    "print 'Regression parameters:', reg.l1_ratio_, reg.alpha_\n",
    "coefs = pd.DataFrame(zip(*[abs(reg.coef_),reg.coef_]), index = transformedParams_test.columns, columns=['abs','true'])\n",
    "print 'Number of features used:', len(coefs) - sum(coefs['abs'] < .00000000001)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#can save state for reproducing estimators later\n",
    "#the pickling of the scikit-learn estimators/regressors will allow the model to be reloaded for prediction of other guide designs, \n",
    "#   but will not be compatible across scikit-learn versions, so it is important to preserve the training data and training/test folds\n",
    "import cPickle\n",
    "estimatorString = cPickle.dumps((fitTable, estimators, scaler, reg))\n",
    "with open(PICKLE_FILE,'w') as outfile:\n",
    "    outfile.write(estimatorString)\n",
    "    \n",
    "#also save the transformed parameters as these can slightly differ based on the automated binning strategy\n",
    "transformedParams_train.head().to_csv(TRANSFORMED_PARAM_HEADER,sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# 2. Applying machine learning model to predict sgRNA activity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate TSS annotation using FANTOM dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#you can supply any table of gene transcription start sites formatted as below\n",
    "#for demonstration purposes, the rest of this walkthrough will use a small arbitrary subset of the protein coding TSS table\n",
    "tssTable_new = tssTable.iloc[10:20, :-1]\n",
    "tssTable_new.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#if desired, use the ensembl annotation and the HGNC database to supply gene aliases to assist P1P2 matching in the next step\n",
    "gencodeData = loadGencodeData('large_data_files/gencode.v19.annotation.gtf')\n",
    "geneToAliases = generateAliasDict('large_data_files/20150424_HGNC_symbols.txt',gencodeData)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Now create a TSS annotation by searching for P1 and P2 peaks near annotated TSSs\n",
    "#same parameters as for our lncRNA libraries\n",
    "p1p2Table_new = generateTssTable_P1P2strategy(tssTable_new, 'large_data_files/TSS_human.sorted.bed.gz', \n",
    "                                                  matchedp1p2Window = 30000, #region around supplied TSS annotation to search for a FANTOM P1 or P2 peak that matches the gene name (or alias)\n",
    "                                                  anyp1p2Window = 500, #region around supplied TSS annotation to search for the nearest P1 or P2 peak\n",
    "                                                  anyPeakWindow = 200, #region around supplied TSS annotation to search for any CAGE peak\n",
    "                                                  minDistanceForTwoTSS = 1000, #If a P1 and P2 peak are found, maximum distance at which to combine into a single annotation (with primary/secondary TSS positions)\n",
    "                                                  aliasDict = geneToAliases[0])\n",
    "#the function will report some collisions of IDs due to use of aliases and redundancy in genome, but will resolve these itself"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p1p2Table_new.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p1p2Table_new.groupby('TSS source').agg(len).iloc[:,[2]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(p1p2Table_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save tables\n",
    "tssTable_alllncs.to_csv(TSS_TABLE_PATH,sep='\\t', index_col=range(2))\n",
    "p1p2Table_alllncs.to_csv(P1P2_TABLE_PATH,sep='\\t', header=0, index_col=range(2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find all sgRNAs in genomic regions of interest "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "libraryTable_new, sgInfoTable_new = findAllGuides(p1p2Table_new, genomeDict, \n",
    "                                                  (-25,500)) #region around P1P2 TSSs to search for new sgRNAs; recommend -550,-25 for CRISPRa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(libraryTable_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "libraryTable_new.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Predicting sgRNA activity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculate parameters for new sgRNAs\n",
    "paramTable_new = generateTypicalParamTable(libraryTable_new, sgInfoTable_new, tssTable_new, p1p2Table_new, genomeDict, bwhandleDict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "paramTable_new.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#if starting from a separate session from where you ran the sgRNA learning steps of Part 1, reload the following\n",
    "import cPickle\n",
    "with open(PICKLE_FILE) as infile:\n",
    "    fitTable, estimators, scaler, reg = cPickle.load(infile)\n",
    "    \n",
    "transformedParams_train = pd.read_csv(TRANSFORMED_PARAM_HEADER,sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#transform and predict scores according to sgRNA prediction model\n",
    "transformedParams_new = transformParams(paramTable_new, fitTable, estimators)\n",
    "\n",
    "#reconcile any differences in column headers generated by automated binning\n",
    "colTups = []\n",
    "for (l1, l2), col in transformedParams_new.iteritems():\n",
    "    colTups.append((l1,str(l2)))\n",
    "transformedParams_new.columns = pd.MultiIndex.from_tuples(colTups)\n",
    "\n",
    "predictedScores_new = pd.Series(reg.predict(scaler.transform(transformedParams_new.loc[:, transformedParams_train.columns].fillna(0).values)), index=transformedParams_new.index)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "predictedScores_new.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "libraryTable_new.to_csv(LIBRARY_TABLE_PATH,sep='\\t')\n",
    "sgInfoTable_new.to_csv(sgRNA_INFO_PATH,sep='\\t')\n",
    "predictedScores_new.to_csv(PREDICTED_SCORES_PATH, sep='\\t')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Construct sgRNA libraries\n",
    "## Score sgRNAs for off-target potential"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### There are many ways to score sgRNAs as off-target; below is one listed one method that is simple and flexible, but ignores gapped alignments, alternate PAMs, and uses bowtie which may not be maximally sensitive in all cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir temp_bowtie_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#output all sequences to a temporary FASTQ file for running bowtie alignment\n",
    "fqFile = 'temp_bowtie_files/bowtie_input.fq'\n",
    "\n",
    "def outputTempBowtieFastq(libraryTable, outputFileName):\n",
    "    phredString = 'I4!=======44444+++++++' #weighting for how impactful mismatches are along sgRNA sequence \n",
    "    with open(outputFileName,'w') as outfile:\n",
    "        for name, row in libraryTable.iterrows():\n",
    "            outfile.write('@' + name + '\\n')\n",
    "            outfile.write('CCN' + str(Seq.Seq(row['sequence'][1:]).reverse_complement()) + '\\n')\n",
    "            outfile.write('+\\n')\n",
    "            outfile.write(phredString + '\\n')\n",
    "            \n",
    "outputTempBowtieFastq(libraryTable_new, fqFile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "#specifying a list of parameters to run bowtie with\n",
    "#each tuple contains\n",
    "# *the mismatch threshold below which a site is considered a potential off-target (higher is more stringent)\n",
    "# *the number of sites allowed (1 is minimum since each sgRNA should have one true site in genome)\n",
    "# *the genome index against which to align the sgRNA sequences; these can be custom built to only consider sites near TSSs\n",
    "# *a name for the bowtie run to create appropriately named output files\n",
    "alignmentList = [(39,1,'large_data_files/hg19.ensemblTSSflank500b','39_nearTSS'),\n",
    "                (31,1,'large_data_files/hg19.ensemblTSSflank500b','31_nearTSS'),\n",
    "                (21,1,'large_data_files/hg19_maskChrMandPAR','21_genome'),\n",
    "                (31,2,'large_data_files/hg19.ensemblTSSflank500b','31_2_nearTSS'),\n",
    "                (31,3,'large_data_files/hg19.ensemblTSSflank500b','31_3_nearTSS')]\n",
    "\n",
    "alignmentColumns = []\n",
    "for btThreshold, mflag, bowtieIndex, runname in alignmentList:\n",
    "\n",
    "    alignedFile = 'temp_bowtie_files/' + runname + '_aligned.txt'\n",
    "    unalignedFile = 'temp_bowtie_files/' + runname + '_unaligned.fq'\n",
    "    maxFile = 'temp_bowtie_files/' + runname + '_max.fq'\n",
    "    \n",
    "    bowtieString = 'bowtie -n 3 -l 15 -e '+str(btThreshold)+' -m ' + str(mflag) + ' --nomaqround -a --tryhard -p 16 --chunkmbs 256 ' + bowtieIndex + ' --suppress 5,6,7 --un ' + unalignedFile + ' --max ' + maxFile + ' '+ ' -q '+fqFile+' '+ alignedFile\n",
    "    print bowtieString\n",
    "    print subprocess.call(bowtieString, shell=True) #0 means finished without errors\n",
    "\n",
    "    #parse through the file of sgRNAs that exceeded \"m\", the maximum allowable alignments, and mark \"True\" any that are found\n",
    "    try:\n",
    "        with open(maxFile) as infile:\n",
    "            sgsAligning = set()\n",
    "            for i, line in enumerate(infile):\n",
    "                if i%4 == 0: #id line\n",
    "                    sgsAligning.add(line.strip()[1:])\n",
    "    except IOError: #no sgRNAs exceeded m, so no maxFile created\n",
    "        sgsAligning = set()\n",
    "                    \n",
    "    alignmentColumns.append(libraryTable_new.apply(lambda row: row.name in sgsAligning, axis=1))\n",
    "    \n",
    "#collate results into a table, and flip the boolean values to yield the sgRNAs that passed filter as True\n",
    "alignmentTable = pd.concat(alignmentColumns,axis=1, keys=zip(*alignmentList)[3]).ne(True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "alignmentTable.head() #True = passed threshold"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pick the top sgRNAs for a library, given predicted activity scores and off-target filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#combine all generated data into one master table\n",
    "predictedScores_new.name = 'predicted score'\n",
    "v2Table = pd.concat((libraryTable_new, predictedScores_new, alignmentTable, sgInfoTable_new), axis=1, keys=['library table v2', 'predicted score', 'off-target filters', 'sgRNA info'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "#for our pCRISPRi/a-v2 vector, we append flanking sequences to each sgRNA sequence for cloning and require the oligo to contain\n",
    "#exactly 1 BstXI and BlpI site each for cloning, and exactly 0 SbfI sites for sequencing sample preparation\n",
    "restrictionSites = {re.compile('CCA......TGG'):1,\n",
    "                   re.compile('GCT.AGC'):1,\n",
    "                   re.compile('CCTGCAGG'):0}\n",
    "\n",
    "def matchREsites(sequence, REdict):\n",
    "    seq = sequence.upper()\n",
    "    for resite, numMatchesExpected in restrictionSites.iteritems():\n",
    "        if len(resite.findall(seq)) != numMatchesExpected:\n",
    "            return False\n",
    "        \n",
    "    return True\n",
    "\n",
    "def checkOverlaps(leftPosition, acceptedLeftPositions, nonoverlapMin):\n",
    "    for pos in acceptedLeftPositions:\n",
    "        if abs(pos - leftPosition) < nonoverlapMin:\n",
    "            return False\n",
    "    return True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#flanking sequences\n",
    "upstreamConstant = 'CCACCTTGTTG'\n",
    "downstreamConstant = 'GTTTAAGAGCTAAGCTG'\n",
    "\n",
    "#minimum overlap between two sgRNAs targeting the same TSS\n",
    "nonoverlapMin = 3\n",
    "\n",
    "#number of sgRNAs to pick per gene/TSS\n",
    "sgRNAsToPick = 10\n",
    "\n",
    "#list of off-target filter (or combinations of filters) levels, matching the names in the alignment table above\n",
    "offTargetLevels = [['31_nearTSS', '21_genome'],\n",
    "                  ['31_nearTSS'],\n",
    "                  ['21_genome'],\n",
    "                  ['31_2_nearTSS'],\n",
    "                  ['31_3_nearTSS']]\n",
    "\n",
    "#for each gene/TSS, go through each sgRNA in descending order of predicted score\n",
    "#if an sgRNA passes the restriction site, overlap, and off-target filters, accept it into the library\n",
    "#if the number of sgRNAs accepted is less than sgRNAsToPick, reduce off-target stringency by one and continue\n",
    "v2Groups = v2Table.groupby([('library table v2','gene'),('library table v2','transcripts')])\n",
    "newSgIds = []\n",
    "unfinishedTss = []\n",
    "for (gene, transcript), group in v2Groups:\n",
    "    geneSgIds = []\n",
    "    geneLeftPositions = []\n",
    "    empiricalSgIds = dict()\n",
    "    \n",
    "    stringency = 0\n",
    "    \n",
    "    while len(geneSgIds) < sgRNAsToPick and stringency < len(offTargetLevels):\n",
    "        for sgId_v2, row in group.sort_values(('predicted score','predicted score'), ascending=False).iterrows():\n",
    "            oligoSeq = upstreamConstant + row[('library table v2','sequence')] + downstreamConstant\n",
    "            leftPos = row[('sgRNA info', 'position')] - (23 if row[('sgRNA info', 'strand')] == '-' else 0)\n",
    "            if len(geneSgIds) < sgRNAsToPick and row['off-target filters'].loc[offTargetLevels[stringency]].all() \\\n",
    "                and matchREsites(oligoSeq, restrictionSites) \\\n",
    "                and checkOverlaps(leftPos, geneLeftPositions, nonoverlapMin):\n",
    "                geneSgIds.append((sgId_v2,\n",
    "                                  gene,transcript,\n",
    "                                  row[('library table v2','sequence')], oligoSeq,\n",
    "                                  row[('predicted score','predicted score')], np.nan,\n",
    "                                 stringency))\n",
    "                geneLeftPositions.append(leftPos)\n",
    "                \n",
    "        stringency += 1\n",
    "            \n",
    "    if len(geneSgIds) < sgRNAsToPick:\n",
    "        unfinishedTss.append((gene, transcript)) #if the number of accepted sgRNAs is still less than sgRNAsToPick, discard gene\n",
    "    else:\n",
    "        newSgIds.extend(geneSgIds)\n",
    "        \n",
    "libraryTable_complete = pd.DataFrame(newSgIds, columns = ['sgID', 'gene', 'transcript','protospacer sequence', 'oligo sequence',\n",
    " 'predicted score', 'empirical score', 'off-target stringency']).set_index('sgID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print len(libraryTable_complete)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#number of sgRNAs accepted at each stringency level\n",
    "libraryTable_complete.groupby('off-target stringency').agg(len).iloc[:,0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#number of TSSs with fewer than required number of sgRNAs (and thus not included in the library)\n",
    "print len(unfinishedTss)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "libraryTable_complete.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Design negative controls matching the base composition of the library"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calcluate the base frequency at each position of the sgRNA, then generate random sequences weighted by this frequency\n",
    "def getBaseFrequencies(libraryTable, baseConversion = {'G':0, 'C':1, 'T':2, 'A':3}):\n",
    "    baseArray = np.zeros((len(libraryTable),20))\n",
    "\n",
    "    for i, (index, seq) in enumerate(libraryTable['protospacer sequence'].iteritems()):\n",
    "        for j, char in enumerate(seq.upper()):\n",
    "            baseArray[i,j] = baseConversion[char]\n",
    "\n",
    "    baseTable = pd.DataFrame(baseArray, index = libraryTable.index)\n",
    "    \n",
    "    baseFrequencies = baseTable.apply(lambda col: col.groupby(col).agg(len)).fillna(0) / len(baseTable)\n",
    "    baseFrequencies.index = ['G','C','T','A']\n",
    "    \n",
    "    baseCumulativeFrequencies = baseFrequencies.copy()\n",
    "    baseCumulativeFrequencies.loc['C'] = baseFrequencies.loc['G'] + baseFrequencies.loc['C']\n",
    "    baseCumulativeFrequencies.loc['T'] = baseFrequencies.loc['G'] + baseFrequencies.loc['C'] + baseFrequencies.loc['T']\n",
    "    baseCumulativeFrequencies.loc['A'] = baseFrequencies.loc['G'] + baseFrequencies.loc['C'] + baseFrequencies.loc['T'] + baseFrequencies.loc['A']\n",
    "\n",
    "    return baseFrequencies, baseCumulativeFrequencies\n",
    "\n",
    "def generateRandomSequence(baseCumulativeFrequencies):\n",
    "    randArray = np.random.random(baseCumulativeFrequencies.shape[1])\n",
    "    \n",
    "    seq = []\n",
    "    for i, col in baseCumulativeFrequencies.iteritems():\n",
    "        for base, freq in col.iteritems():\n",
    "            if randArray[i] < freq:\n",
    "                seq.append(base)\n",
    "                break\n",
    "                \n",
    "    return ''.join(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "baseCumulativeFrequencies = getBaseFrequencies(libraryTable_complete)[1]\n",
    "negList = []\n",
    "numberToGenerate = 1000 #can generate many more; some will be filtered out by off-targets, and you can always select an arbitrary subset for inclusion into the library\n",
    "for i in range(numberToGenerate):\n",
    "    negList.append(generateRandomSequence(baseCumulativeFrequencies))\n",
    "negTable = pd.DataFrame(negList, index=['non-targeting_' + str(i) for i in range(numberToGenerate)], columns = ['sequence'])\n",
    "\n",
    "fqFile = 'temp_bowtie_files/bowtie_input_negs.fq'\n",
    "outputTempBowtieFastq(negTable, fqFile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#similar to targeting sgRNA off-target scoring, but looking for sgRNAs with 0 alignments\n",
    "alignmentList = [(31,1,'~/indices/hg19.ensemblTSSflank500b','31_nearTSS_negs'),\n",
    "                (21,1,'~/indices/hg19_maskChrMandPAR','21_genome_negs')]\n",
    "\n",
    "alignmentColumns = []\n",
    "for btThreshold, mflag, bowtieIndex, runname in alignmentList:\n",
    "\n",
    "    alignedFile = 'temp_bowtie_files/' + runname + '_aligned.txt'\n",
    "    unalignedFile = 'temp_bowtie_files/' + runname + '_unaligned.fq'\n",
    "    maxFile = 'temp_bowtie_files/' + runname + '_max.fq'\n",
    "    \n",
    "    bowtieString = 'bowtie -n 3 -l 15 -e '+str(btThreshold)+' -m ' + str(mflag) + ' --nomaqround -a --tryhard -p 16 --chunkmbs 256 ' + bowtieIndex + ' --suppress 5,6,7 --un ' + unalignedFile + ' --max ' + maxFile + ' '+ ' -q '+fqFile+' '+ alignedFile\n",
    "    print bowtieString\n",
    "    print subprocess.call(bowtieString, shell=True)\n",
    "\n",
    "    #read unaligned file for negs, and then don't flip boolean of alignmentTable\n",
    "    with open(unalignedFile) as infile:\n",
    "        sgsAligning = set()\n",
    "        for i, line in enumerate(infile):\n",
    "            if i%4 == 0: #id line\n",
    "                sgsAligning.add(line.strip()[1:])\n",
    "\n",
    "    alignmentColumns.append(negTable.apply(lambda row: row.name in sgsAligning, axis=1))\n",
    "    \n",
    "alignmentTable = pd.concat(alignmentColumns,axis=1, keys=zip(*alignmentList)[3])\n",
    "alignmentTable.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "acceptedNegList = []\n",
    "negCount = 0\n",
    "for i, (name, row) in enumerate(pd.concat((negTable,alignmentTable),axis=1, keys=['seq','alignment']).iterrows()):\n",
    "    oligo = upstreamConstant + row['seq','sequence'] + downstreamConstant\n",
    "    if row['alignment'].all() and matchREsites(oligo, restrictionSites):\n",
    "        acceptedNegList.append(('non-targeting_%05d' % negCount, 'negative_control', 'na', row['seq','sequence'], oligo, 0))\n",
    "        negCount += 1\n",
    "        \n",
    "acceptedNegs = pd.DataFrame(acceptedNegList, columns = ['sgId', 'gene', 'transcript', 'protospacer sequence', 'oligo sequence', 'off-target stringency']).set_index('sgId')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "acceptedNegs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "libraryTable_complete.to_csv(LIBRARY_WITHOUT_NEGATIVES_PATH, sep='\\t')\n",
    "acceptedNegs.to_csv(NEGATIVE_CONTROLS_PATH,sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Finalizing library design"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* divide genes into sublibrary groups (if required)\n",
    "* assign negative control sgRNAs to sublibrary groups; ~1-2% of the number of sgRNAs in the library is a good rule-of-thumb\n",
    "* append PCR adapter sequences (~18bp) to each end of the oligo sequences to enable amplification of the oligo pool; each sublibary should have an orthogonal sequence so they can be cloned separately"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (CRISPRiaDesign)",
   "language": "python",
   "name": "crispriadesign"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
